{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "jupyter nbconvert 5_association.ipynb --to script\n",
    "\n",
    "cd /data/ch6845/MHC*;screen -S assoc;\n",
    "\n",
    "for i in {0..10};do python 4_association.py $i;done\n",
    "\n",
    "\n",
    "5*3=15\n",
    "\n",
    "for i in {00..101};do python 5_association.py $i;done\n",
    "\n",
    "https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6292650/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "\n",
    "import logging\n",
    "import re\n",
    "import pathlib\n",
    "\n",
    "import datetime\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from scipy.stats import pearsonr\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from pyplink import PyPlink\n",
    "\n",
    "import statsmodels.api as sm\n",
    "\n",
    "from basic_tools import *\n",
    "\n",
    "HLA_re_exp=re.compile('(?P<name>HLA_[0-9A-Z]*)\\*(?P<allele>[0-9:]*)')\n",
    "AA_re_exp=re.compile('(?P<name>AA_[A-Z0-9]*_[0-9]*_[0-9]*_exon[0-9]*)_*(?P<allele>[A-Z]*)')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "plink_KCHIP_HLA_AA_SNP_1000G_path=data_genotype_merge_path+'KCHIP_HLA_AA_SNP_1000G'\n",
    "plink_KCHIP_HLA_SNP_1000G_path=data_genotype_merge_path+'KCHIP_HLA_SNP_1000G' # for association\n",
    "plink_1000G_path=data_genotype_merge_path+'1000G'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plink_KCHIP_HLA_SNP_1000G=PyPlink(plink_KCHIP_HLA_SNP_1000G_path)\n",
    "#plink_KCHIP_HLA_SNP_1000G_fam=plink_KCHIP_HLA_SNP_1000G.get_fam().astype({'fid':str,'iid':str}).rename(columns={'fid':'FID','iid':'IID'})\n",
    "#plink_KCHIP_HLA_SNP_1000G_bim=plink_KCHIP_HLA_SNP_1000G.get_bim()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "plink_KCHIP_HLA_AA_SNP_1000G=PyPlink(plink_KCHIP_HLA_AA_SNP_1000G_path)\n",
    "plink_KCHIP_HLA_AA_SNP_1000G_fam=plink_KCHIP_HLA_AA_SNP_1000G.get_fam().astype({'fid':str,'iid':str}).rename(columns={'fid':'FID','iid':'IID'})\n",
    "plink_KCHIP_HLA_AA_SNP_1000G_bim=plink_KCHIP_HLA_AA_SNP_1000G.get_bim()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "plink_1000G=PyPlink(plink_1000G_path)\n",
    "plink_1000G_fam=plink_1000G.get_fam().astype({'fid':str,'iid':str}).rename(columns={'fid':'FID','iid':'IID'})\n",
    "plink_1000G_bim=plink_1000G.get_bim()\n",
    "# plink_KCHIP_HLA_SNP_1000G_bim -> for parameter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'data/genotype/4_merge/KCHIP_HLA_AA_SNP.bgl.phased'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phased_KCHIP_HLA_AA_path\n",
    "phased_KCHIP_HLA_AA_SNP_path"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# parse parameter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "phenotypes=pd.read_csv(pheno_all_file_path,sep='\\t')\n",
    "phenotypes=phenotypes.set_index('ID').loc[plink_KCHIP_HLA_AA_SNP_1000G_fam['IID']]\n",
    "\n",
    "binary_continuous_traits=sorted(phenotypes.columns[~phenotypes.columns.str.contains('x_ray')])\n",
    "#binary_continuous_traits\n",
    "\n",
    "if 'ipykernel' in sys.argv[0]:\n",
    "    ipykernel=True\n",
    "    phenotype_name='ALP'\n",
    "    #phenotype_name='height'\n",
    "else:\n",
    "    ipykernel=False\n",
    "    phenotype_name=sys.argv[1]\n",
    "if phenotype_name.isdigit():\n",
    "    phenotype_name=int(phenotype_name)\n",
    "    phenotype_name=binary_continuous_traits[phenotype_name]      "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "pheno=pd.read_csv(data_out_pheno_path+phenotype_name+'.phe',sep='\\t',names=['FID','IID','pheno'])\n",
    "phenotype_type='binary' if len(pheno['pheno'].value_counts())<3 else 'continuous'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_out_assoc_phenotype_path=data_out_assoc_path+phenotype_name+'/'\n",
    "pathlib.Path(data_out_assoc_phenotype_path).mkdir(parents=True, exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "f=open(phased_KCHIP_HLA_AA_path,'r');aa_ind=f.readline().strip().split(' ')[2:];f.close()\n",
    "\n",
    "aa_ind_1=[aa_ind[i] for i in range(0,len(aa_ind),2)]\n",
    "aa_ind_2=[aa_ind[i+1] for i in range(0,len(aa_ind),2)]\n",
    "\n",
    "assert np.all(pheno['FID']==aa_ind_1)\n",
    "assert np.all(pheno['FID']==aa_ind_2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "log = logging.getLogger('logger')\n",
    "log.setLevel(logging.DEBUG)\n",
    "\n",
    "log_file_name=datetime.datetime.now().strftime('%Y%m%d_%H%M%S')+'.log'\n",
    "log_file_path=data_out_assoc_phenotype_path+log_file_name\n",
    "fileHandler = logging.FileHandler(log_file_path)\n",
    "streamHandler = logging.StreamHandler()\n",
    "\n",
    "formatter = logging.Formatter('> %(message)s')\n",
    "fileHandler.setFormatter(formatter)\n",
    "streamHandler.setFormatter(formatter)\n",
    "\n",
    "log.addHandler(fileHandler)\n",
    "log.addHandler(streamHandler)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "log.info_head=lambda x: log.info('-'*int((100-len(x))/2)+x+'-'*int((100-len(x))/2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "> ---------------------------phenotype_name: ALP, phenotype_type:continuous---------------------------\n"
     ]
    }
   ],
   "source": [
    "log.info_head(\"phenotype_name: {}, phenotype_type:{}\".format(phenotype_name,phenotype_type))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Association"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "plink_KCHIP_HLA_AA_SNP_1000G_HLAassign=plink_KCHIP_HLA_AA_SNP_1000G_bim[['pos']]\n",
    "HLA_names=np.unique([i[0].split('_')[1] for i in plink_KCHIP_HLA_AA_SNP_1000G_bim[plink_KCHIP_HLA_AA_SNP_1000G_bim.index.str.contains('HLA_')].index.str.split('*')])\n",
    "\n",
    "for HLA_name in HLA_names:\n",
    "    plink_KCHIP_HLA_AA_SNP_1000G_HLAassign[HLA_name]=0\n",
    "    gene_select=plink_KCHIP_HLA_AA_SNP_1000G_HLAassign[plink_KCHIP_HLA_AA_SNP_1000G_HLAassign.index.str.contains('HLA_'+HLA_name)|plink_KCHIP_HLA_AA_SNP_1000G_HLAassign.index.str.contains('SNPS_'+HLA_name)|plink_KCHIP_HLA_AA_SNP_1000G_HLAassign.index.str.contains('AA_'+HLA_name)]#print(gene_select.sort_values('pos').iloc[0],gene_select.sort_values('pos').iloc[-1])\n",
    "    plink_KCHIP_HLA_AA_SNP_1000G_HLAassign[HLA_name][(plink_KCHIP_HLA_AA_SNP_1000G_HLAassign['pos']>=plink_KCHIP_HLA_AA_SNP_1000G_HLAassign['pos'].min())&(plink_KCHIP_HLA_AA_SNP_1000G_HLAassign['pos']<=plink_KCHIP_HLA_AA_SNP_1000G_HLAassign['pos'].max())]=1\n",
    "plink_KCHIP_HLA_AA_SNP_1000G_HLAassign"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "#for i in gene_assign[gene_assign.A==1].index:\n",
    "#    print(i,gene_assign.loc[i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ch6845/tools/miniconda3/envs/pytorch/lib/python3.6/site-packages/ipykernel_launcher.py:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  \"\"\"\n",
      "/home/ch6845/tools/miniconda3/envs/pytorch/lib/python3.6/site-packages/pandas/core/generic.py:9114: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self._update_inplace(new_data)\n",
      "/home/ch6845/tools/miniconda3/envs/pytorch/lib/python3.6/site-packages/IPython/core/interactiveshell.py:3319: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  exec(code_obj, self.user_global_ns, self.user_ns)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos</th>\n",
       "      <th>A</th>\n",
       "      <th>B</th>\n",
       "      <th>C</th>\n",
       "      <th>DPA1</th>\n",
       "      <th>DPB1</th>\n",
       "      <th>DQA1</th>\n",
       "      <th>DQB1</th>\n",
       "      <th>DRB1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>snp</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6:28000361_T/C</th>\n",
       "      <td>28000361</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6:28000610_C/A</th>\n",
       "      <td>28000610</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6:28001003_A/C</th>\n",
       "      <td>28001003</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6:28001033_T/C</th>\n",
       "      <td>28001033</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6:28001610_T/C</th>\n",
       "      <td>28001610</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6:34997188_C/T</th>\n",
       "      <td>34997188</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6:34997259_G/A</th>\n",
       "      <td>34997259</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6:34997606_T/A</th>\n",
       "      <td>34997606</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6:34998418_G/A</th>\n",
       "      <td>34998418</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6:34999077_C/T</th>\n",
       "      <td>34999077</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>74907 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     pos  A  B  C  DPA1  DPB1  DQA1  DQB1  DRB1\n",
       "snp                                                            \n",
       "6:28000361_T/C  28000361  0  0  0     0     0     0     0     0\n",
       "6:28000610_C/A  28000610  0  0  0     0     0     0     0     0\n",
       "6:28001003_A/C  28001003  0  0  0     0     0     0     0     0\n",
       "6:28001033_T/C  28001033  0  0  0     0     0     0     0     0\n",
       "6:28001610_T/C  28001610  0  0  0     0     0     0     0     0\n",
       "...                  ... .. .. ..   ...   ...   ...   ...   ...\n",
       "6:34997188_C/T  34997188  0  0  0     0     0     0     0     0\n",
       "6:34997259_G/A  34997259  0  0  0     0     0     0     0     0\n",
       "6:34997606_T/A  34997606  0  0  0     0     0     0     0     0\n",
       "6:34998418_G/A  34998418  0  0  0     0     0     0     0     0\n",
       "6:34999077_C/T  34999077  0  0  0     0     0     0     0     0\n",
       "\n",
       "[74907 rows x 9 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "plink_KCHIP_HLA_AA_SNP_1000G_HLAassign=plink_KCHIP_HLA_AA_SNP_1000G_bim[['pos']]\n",
    "HLA_names=np.unique([i[0].split('_')[1] for i in plink_KCHIP_HLA_AA_SNP_1000G_bim[plink_KCHIP_HLA_AA_SNP_1000G_bim.index.str.contains('HLA_')].index.str.split('*')])\n",
    "\n",
    "for HLA_name in HLA_names:\n",
    "    plink_KCHIP_HLA_AA_SNP_1000G_HLAassign[HLA_name]=0\n",
    "    gene_select=plink_KCHIP_HLA_AA_SNP_1000G_HLAassign[plink_KCHIP_HLA_AA_SNP_1000G_HLAassign.index.str.contains('HLA_'+HLA_name)|plink_KCHIP_HLA_AA_SNP_1000G_HLAassign.index.str.contains('SNPS_'+HLA_name)|plink_KCHIP_HLA_AA_SNP_1000G_HLAassign.index.str.contains('AA_'+HLA_name)]#print(gene_select.sort_values('pos').iloc[0],gene_select.sort_values('pos').iloc[-1])\n",
    "    plink_KCHIP_HLA_AA_SNP_1000G_HLAassign[HLA_name][(plink_KCHIP_HLA_AA_SNP_1000G_HLAassign['pos']>=gene_select['pos'].min())&(plink_KCHIP_HLA_AA_SNP_1000G_HLAassign['pos']<=gene_select['pos'].max())]=1\n",
    "plink_KCHIP_HLA_AA_SNP_1000G_HLAassign"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "plink_KCHIP_HLA_AA_SNP_1000G_r2_list_dict={}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "conditional_variant_list=['AA_A_-1','SNPS_A_28_30018337_exon1','6:30165273_C/T']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'data/genotype/4_merge/KCHIP_HLA_AA.bgl.phased'"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phased_KCHIP_HLA_AA_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Initialize *.aa\n",
    "idx=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "log.info_head(\"Step {}-1 Initialization\".format(idx))\n",
    "\n",
    "if os.path.exists('{}'.format(data_out_assoc_phenotype_path+'step_{:02d}.bgl.phased'.format(idx))):\n",
    "    command='rm {}'.format(data_out_assoc_phenotype_path+'step_{:02d}.*'.format(idx))\n",
    "    log.info(command);os.system(command)\n",
    "\n",
    "\n",
    "command='cp {} {}'.format(phased_KCHIP_HLA_AA_path,data_out_assoc_phenotype_path+'step_{:02d}.bgl.phased'.format(idx))\n",
    "log.info(command); assert os.system(command)==0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "> ------------------------------Step 1-2 Classifying conditional variant------------------------------\n",
      "> conditional_variant: AA_A_-1\n",
      "> added to conditional_variant_list_HLA_AA\n",
      "> conditional_variant: SNPS_A_28_30018337_exon1\n",
      "> added to conditional_variant_list_HLA_AA\n",
      "> added to conditional_variant_list_phased_SNP\n",
      "> conditional_variant: 6:30165273_C/T\n",
      "> added to conditional_variant_list_unphased_SNP\n",
      "> calculating r2\n",
      "> r2 result:  \n",
      "A       0\n",
      "B       0\n",
      "C       0\n",
      "DPA1    0\n",
      "DPB1    0\n",
      "DQA1    0\n",
      "DQB1    0\n",
      "DRB1    0\n",
      "dtype: int64\n",
      "> conditional_variant_list_HLA_AA ['A']\n",
      "> conditional_variant_list_phased_SNP ['SNPS_A_28_30018337_exon1']\n",
      "> conditional_variant_list_unphased_SNP ['6:30165273_C/T']\n"
     ]
    }
   ],
   "source": [
    "log.info_head(\"Step {}-2 Classifying conditional variant\".format(idx))\n",
    "\n",
    "conditional_variant_list_HLA_AA=[]\n",
    "conditional_variant_list_phased_SNP=[]\n",
    "conditional_variant_list_unphased_SNP=[]\n",
    "\n",
    "\n",
    "for conditional_variant in conditional_variant_list:\n",
    "    log.info(\"conditional_variant: {}\".format(conditional_variant))\n",
    "    if conditional_variant[:3]=='AA_':\n",
    "        log.info(\"added to conditional_variant_list_HLA_AA\")\n",
    "        conditional_variant_list_HLA_AA.append(conditional_variant.split('_')[1])\n",
    "    elif conditional_variant[:4]=='HLA_':\n",
    "        log.info(\"added to conditional_variant_list_HLA_AA\")\n",
    "        conditional_variant_list_HLA_AA.append(conditional_variant[4:].split('*')[0])\n",
    "    elif conditional_variant[:5]=='SNPS_':\n",
    "        log.info(\"added to conditional_variant_list_HLA_AA\")\n",
    "        log.info(\"added to conditional_variant_list_phased_SNP\")\n",
    "        conditional_variant_list_HLA_AA.append(conditional_variant.split('_')[1])\n",
    "        conditional_variant_list_phased_SNP.append(conditional_variant)\n",
    "    else:\n",
    "        if conditional_variant in plink_1000G_bim.index:\n",
    "            log.info(\"added to conditional_variant_list_unphased_SNP\")\n",
    "            conditional_variant_list_unphased_SNP.append(conditional_variant)\n",
    "        else:\n",
    "            log.info(\"added to conditional_variant_list_phased_SNP\")\n",
    "            conditional_variant_list_phased_SNP.append(conditional_variant)\n",
    "        log.info(\"calculating r2\")\n",
    "        r2_list=plink_KCHIP_HLA_AA_SNP_1000G_r2_list_dict.get(conditional_variant)\n",
    "        if r2_list is None: \n",
    "            r2_list=[]\n",
    "            for idx_bim,(SNP,row) in enumerate(plink_KCHIP_HLA_AA_SNP_1000G_bim.iterrows()):\n",
    "                r2=pearsonr(plink_KCHIP_HLA_AA_SNP_1000G.get_geno_marker(conditional_variant),plink_KCHIP_HLA_AA_SNP_1000G.get_geno_marker(SNP))[0]**2\n",
    "                r2_list.append(r2)\n",
    "            plink_KCHIP_HLA_AA_SNP_1000G_r2_list_dict[conditional_variant]=r2_list\n",
    "        r2_df=pd.DataFrame(r2_list,index=plink_KCHIP_HLA_AA_SNP_1000G_bim.index)             \n",
    "        plink_KCHIP_HLA_AA_SNP_1000G_HLAassign_count=plink_KCHIP_HLA_AA_SNP_1000G_HLAassign[r2_df[0]>0.7].sum(axis=0).drop('pos')\n",
    "        log.info(\"r2 result:  \\n{}\".format(plink_KCHIP_HLA_AA_SNP_1000G_HLAassign_count))\n",
    "        for HLA_name in plink_KCHIP_HLA_AA_SNP_1000G_HLAassign_count[plink_KCHIP_HLA_AA_SNP_1000G_HLAassign_count>0].index.tolist():\n",
    "            log.info(\"{} added to conditional_variant_list_HLA_AA\".format(HLA_name))\n",
    "            conditional_variant_list_HLA_AA.append(HLA_name)\n",
    "\n",
    "\n",
    "conditional_variant_list_HLA_AA=np.unique(conditional_variant_list_HLA_AA)\n",
    "conditional_variant_list_phased_SNP=np.unique(conditional_variant_list_phased_SNP)\n",
    "conditional_variant_list_unphased_SNP=np.unique(conditional_variant_list_unphased_SNP)\n",
    "\n",
    "\n",
    "log.info(\"conditional_variant_list_HLA_AA {}\".format(conditional_variant_list_HLA_AA))\n",
    "log.info(\"conditional_variant_list_phased_SNP {}\".format(conditional_variant_list_phased_SNP))\n",
    "log.info(\"conditional_variant_list_unphased_SNP {}\".format(conditional_variant_list_unphased_SNP))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n"
     ]
    }
   ],
   "source": [
    "print(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "> ------------------------Step 1-3 Constructing covariate and conditional list------------------------\n",
      "> conditional_plink_list 0 : []\n",
      "> -----------\n",
      "> covariate_plink_df 104 : Index(['PC1', 'PC2', 'PC3', 'PC4', 'age', 'sex', 'AS', 'NC', 'HLA_A*01:01',\n",
      "       'HLA_A*02:01',\n",
      "       ...\n",
      "       'AA_A_294_30020314_exon5', 'AA_A_297_30020323_exon5',\n",
      "       'AA_A_298_30020326_exon5', 'AA_A_299_30020329_exon5',\n",
      "       'AA_A_307_30020353_exon5', 'AA_A_311_30020365_exon5',\n",
      "       'AA_A_321_30020837_exon6', 'AA_A_334_30021018_exon7',\n",
      "       'SNPS_A_28_30018337_exon1', '6:30165273_C/T'],\n",
      "      dtype='object', length=104)\n",
      "> -----------\n",
      "> conditional_omnibus_list 61 : ['HLA_A', 'AA_A_9_30018537_exon2', 'AA_A_17_30018561_exon2', 'AA_A_44_30018642_exon2', 'AA_A_56_30018678_exon2', 'AA_A_62_30018696_exon2', 'AA_A_63_30018699_exon2', 'AA_A_65_30018705_exon2', 'AA_A_66_30018708_exon2', 'AA_A_67_30018711_exon2', 'AA_A_70_30018720_exon2', 'AA_A_73_30018729_exon2', 'AA_A_74_30018732_exon2', 'AA_A_76_30018738_exon2', 'AA_A_77_30018741_exon2', 'AA_A_79_30018747_exon2', 'AA_A_80_30018750_exon2', 'AA_A_81_30018753_exon2', 'AA_A_82_30018756_exon2', 'AA_A_83_30018759_exon2', 'AA_A_90_30018780_exon2', 'AA_A_95_30019036_exon3', 'AA_A_97_30019042_exon3', 'AA_A_99_30019048_exon3', 'AA_A_102_30019057_exon3', 'AA_A_105_30019066_exon3', 'AA_A_107_30019072_exon3', 'AA_A_109_30019078_exon3', 'AA_A_114_30019093_exon3', 'AA_A_116_30019099_exon3', 'AA_A_127_30019132_exon3', 'AA_A_142_30019177_exon3', 'AA_A_144_30019183_exon3', 'AA_A_145_30019186_exon3', 'AA_A_149_30019198_exon3', 'AA_A_150_30019201_exon3', 'AA_A_151_30019204_exon3', 'AA_A_152_30019207_exon3', 'AA_A_156_30019219_exon3', 'AA_A_158_30019225_exon3', 'AA_A_161_30019234_exon3', 'AA_A_163_30019240_exon3', 'AA_A_166_30019249_exon3', 'AA_A_167_30019252_exon3', 'AA_A_184_30019882_exon4', 'AA_A_193_30019909_exon4', 'AA_A_194_30019912_exon4', 'AA_A_207_30019951_exon4', 'AA_A_246_30020068_exon4', 'AA_A_253_30020089_exon4', 'AA_A_276_30020260_exon5', 'AA_A_282_30020278_exon5', 'AA_A_294_30020314_exon5', 'AA_A_297_30020323_exon5', 'AA_A_298_30020326_exon5', 'AA_A_299_30020329_exon5', 'AA_A_307_30020353_exon5', 'AA_A_311_30020365_exon5', 'AA_A_321_30020837_exon6', 'AA_A_334_30021018_exon7', 'SNPS_A_28_30018337_exon1']\n",
      "> -----------\n",
      "> covariate_omnibus_df 9 : Index(['PC1', 'PC2', 'PC3', 'PC4', 'age', 'sex', 'AS', 'NC', '6:30165273_C/T'], dtype='object')\n"
     ]
    }
   ],
   "source": [
    "    log.info_head(\"Step {}-3 Constructing covariate and conditional list\".format(idx))\n",
    "\n",
    "    covariate_df=pd.read_csv(PC_path,sep='\\t').set_index('ID').loc[plink_KCHIP_HLA_AA_SNP_1000G_fam['IID']]\n",
    "    covariate_df['age']=phenotypes['age']\n",
    "    covariate_df['sex']=phenotypes['sex']-1\n",
    "    covariate_df['AS']=phenotypes['cohort'].replace(1,1).replace(2,0).replace(3,0)\n",
    "    covariate_df['NC']=phenotypes['cohort'].replace(1,0).replace(2,0).replace(3,1)\n",
    "\n",
    "    covariate_plink_df=covariate_df.copy()\n",
    "    covariate_omnibus_df=covariate_df.copy()\n",
    "\n",
    "    conditional_plink_list=[]\n",
    "    conditional_omnibus_list=[]\n",
    "\n",
    "    for HLA_gene in conditional_variant_list_HLA_AA:\n",
    "        marker_dosage_list_dict={}\n",
    "        for re_exp in [HLA_re_exp,AA_re_exp]:\n",
    "            for marker in plink_KCHIP_HLA_AA_SNP_1000G_bim.index:\n",
    "                name,allele=(re_exp.search(marker).group('name'),re_exp.search(marker).group('allele')) if re_exp.search(marker) is not None else (None,None)\n",
    "                if name is not None and name.split('*')[0].split('_')[1]==HLA_gene:\n",
    "                    marker_dosage_list=marker_dosage_list_dict.get(name,[])\n",
    "                    marker_dosage_list.append({\"marker\":marker,\"dosage\":plink_KCHIP_HLA_AA_SNP_1000G.get_geno_marker(marker)})\n",
    "                    marker_dosage_list_dict[name]=marker_dosage_list\n",
    "\n",
    "        for key,marker_dosage_list in marker_dosage_list_dict.items():\n",
    "            marker_list=[marker_dosage['marker'] for marker_dosage in marker_dosage_list]\n",
    "            dosage_array=np.array([marker_dosage['dosage'] for marker_dosage in marker_dosage_list]).transpose()\n",
    "            if len(marker_list)>1:\n",
    "                marker_list_cut=np.delete(marker_list,dosage_array.sum(axis=0).argmax())\n",
    "                dosage_array_cut=np.delete(dosage_array,dosage_array.sum(axis=0).argmax(),axis=1)\n",
    "            else:\n",
    "                marker_list_cut=marker_list\n",
    "                dosage_array_cut=dosage_array\n",
    "\n",
    "            for i,marker in enumerate(marker_list_cut):\n",
    "                covariate_plink_df[marker]=dosage_array_cut[:,i]\n",
    "\n",
    "        conditional_omnibus_list+=marker_dosage_list_dict.keys()\n",
    "\n",
    "\n",
    "    for phased_SNP in conditional_variant_list_phased_SNP:\n",
    "        dosage=plink_KCHIP_HLA_AA_SNP_1000G.get_geno_marker(phased_SNP).astype(float)\n",
    "        dosage[dosage==-1]=np.nan\n",
    "\n",
    "        covariate_plink_df[phased_SNP]=dosage\n",
    "        #conditional_plink_list.append(phased_SNP)\n",
    "        conditional_omnibus_list.append(phased_SNP)\n",
    "\n",
    "        assert os.system('grep \"{} \" {} > {}temp.bgl.phased'.format(phased_SNP,phased_KCHIP_HLA_AA_SNP_path,data_out_assoc_phenotype_path))==0\n",
    "        f=open('{}temp.bgl.phased'.format(data_out_assoc_phenotype_path),'r');lines=f.readlines();f.close();assert len(lines)==1\n",
    "        assert os.system('cat {}temp.bgl.phased >> {}'.format(data_out_assoc_phenotype_path,data_out_assoc_phenotype_path+'step_{:02d}.bgl.phased'.format(idx)))==0\n",
    "        assert os.system('rm {}temp.bgl.phased'.format(data_out_assoc_phenotype_path))==0    \n",
    "\n",
    "\n",
    "    for unphased_SNP in conditional_variant_list_unphased_SNP:\n",
    "        dosage=plink_KCHIP_HLA_AA_SNP_1000G.get_geno_marker(unphased_SNP).astype(float)\n",
    "        dosage[dosage==-1]=np.nan\n",
    "        covariate_plink_df[unphased_SNP]=dosage\n",
    "        covariate_omnibus_df[unphased_SNP]=dosage\n",
    "\n",
    "    log.info(\"conditional_plink_list {} : {}\".format(len(conditional_plink_list),conditional_plink_list))\n",
    "    log.info(\"-----------\")\n",
    "    log.info(\"covariate_plink_df {} : {}\".format(len(covariate_plink_df.columns),covariate_plink_df.columns))         \n",
    "    log.info(\"-----------\")\n",
    "    log.info(\"conditional_omnibus_list {} : {}\".format(len(conditional_omnibus_list),conditional_omnibus_list))\n",
    "    log.info(\"-----------\")\n",
    "    log.info(\"covariate_omnibus_df {} : {}\".format(len(covariate_omnibus_df.columns),covariate_omnibus_df.columns))    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [],
   "source": [
    "    def check_duplicate(array,axis=1,threshold=0.9,keep='last'):\n",
    "        array=array.copy()\n",
    "        if isinstance(array, np.ndarray):\n",
    "            pass\n",
    "        elif isinstance(array, pd.core.frame.DataFrame):\n",
    "            array=array.values\n",
    "\n",
    "        if axis==1:\n",
    "            pass\n",
    "        elif axis==0:\n",
    "            array=array.transpose()\n",
    "        else:\n",
    "            raise NotImplementedError\n",
    "\n",
    "        dupcol_check=np.repeat(False,array.shape[1])\n",
    "        for i in range(array.shape[1]):\n",
    "            for j in range(i+1,array.shape[1]):\n",
    "                x_i=array[:,i]\n",
    "                x_j=array[:,j]\n",
    "                if np.all(x_i==x_j) or np.corrcoef(x_i,x_j)[0,1]>threshold:\n",
    "                    if keep==False:\n",
    "                        dupcol_check[i]=True\n",
    "                        dupcol_check[j]=True\n",
    "                    elif keep=='last':   \n",
    "                        if (not dupcol_check[i]) and (not dupcol_check[j]):\n",
    "                            dupcol_check[j]=True\n",
    "                    elif keep=='loose_last':\n",
    "                        dupcol_check[j]=True\n",
    "                    else:\n",
    "                        raise NotImplementedError\n",
    "        return dupcol_check"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "> -----------------------------------Step 1-4 Checking colinearity-----------------------------------\n",
      "> conditional_plink_list 0 : []\n",
      "> -----------\n",
      "> covariate_plink_df 104 : Index(['PC1', 'PC2', 'PC3', 'PC4', 'age', 'sex', 'AS', 'NC', 'HLA_A*01:01',\n",
      "       'HLA_A*02:01',\n",
      "       ...\n",
      "       'AA_A_294_30020314_exon5', 'AA_A_297_30020323_exon5',\n",
      "       'AA_A_298_30020326_exon5', 'AA_A_299_30020329_exon5',\n",
      "       'AA_A_307_30020353_exon5', 'AA_A_311_30020365_exon5',\n",
      "       'AA_A_321_30020837_exon6', 'AA_A_334_30021018_exon7',\n",
      "       'SNPS_A_28_30018337_exon1', '6:30165273_C/T'],\n",
      "      dtype='object', length=104)\n",
      "> -----------\n",
      "> conditional_omnibus_list 61 : ['HLA_A', 'AA_A_9_30018537_exon2', 'AA_A_17_30018561_exon2', 'AA_A_44_30018642_exon2', 'AA_A_56_30018678_exon2', 'AA_A_62_30018696_exon2', 'AA_A_63_30018699_exon2', 'AA_A_65_30018705_exon2', 'AA_A_66_30018708_exon2', 'AA_A_67_30018711_exon2', 'AA_A_70_30018720_exon2', 'AA_A_73_30018729_exon2', 'AA_A_74_30018732_exon2', 'AA_A_76_30018738_exon2', 'AA_A_77_30018741_exon2', 'AA_A_79_30018747_exon2', 'AA_A_80_30018750_exon2', 'AA_A_81_30018753_exon2', 'AA_A_82_30018756_exon2', 'AA_A_83_30018759_exon2', 'AA_A_90_30018780_exon2', 'AA_A_95_30019036_exon3', 'AA_A_97_30019042_exon3', 'AA_A_99_30019048_exon3', 'AA_A_102_30019057_exon3', 'AA_A_105_30019066_exon3', 'AA_A_107_30019072_exon3', 'AA_A_109_30019078_exon3', 'AA_A_114_30019093_exon3', 'AA_A_116_30019099_exon3', 'AA_A_127_30019132_exon3', 'AA_A_142_30019177_exon3', 'AA_A_144_30019183_exon3', 'AA_A_145_30019186_exon3', 'AA_A_149_30019198_exon3', 'AA_A_150_30019201_exon3', 'AA_A_151_30019204_exon3', 'AA_A_152_30019207_exon3', 'AA_A_156_30019219_exon3', 'AA_A_158_30019225_exon3', 'AA_A_161_30019234_exon3', 'AA_A_163_30019240_exon3', 'AA_A_166_30019249_exon3', 'AA_A_167_30019252_exon3', 'AA_A_184_30019882_exon4', 'AA_A_193_30019909_exon4', 'AA_A_194_30019912_exon4', 'AA_A_207_30019951_exon4', 'AA_A_246_30020068_exon4', 'AA_A_253_30020089_exon4', 'AA_A_276_30020260_exon5', 'AA_A_282_30020278_exon5', 'AA_A_294_30020314_exon5', 'AA_A_297_30020323_exon5', 'AA_A_298_30020326_exon5', 'AA_A_299_30020329_exon5', 'AA_A_307_30020353_exon5', 'AA_A_311_30020365_exon5', 'AA_A_321_30020837_exon6', 'AA_A_334_30021018_exon7', 'SNPS_A_28_30018337_exon1']\n",
      "> -----------\n",
      "> covariate_omnibus_df 9 : Index(['PC1', 'PC2', 'PC3', 'PC4', 'age', 'sex', 'AS', 'NC', '6:30165273_C/T'], dtype='object')\n"
     ]
    }
   ],
   "source": [
    "    log.info_head(\"Step {}-4 Checking colinearity\".format(idx))\n",
    "    covariate_plink_df_drop=covariate_plink_df#.drop(columns=covariate_plink_df.columns[check_duplicate(covariate_plink_df,threshold=0.8)])\n",
    "    covariate_omnibus_df_drop=covariate_omnibus_df#.drop(columns=covariate_omnibus_df.columns[check_duplicate(covariate_omnibus_df,threshold=0.9)])\n",
    "    \n",
    "    log.info(\"conditional_plink_list {} : {}\".format(len(conditional_plink_list),conditional_plink_list))\n",
    "    log.info(\"-----------\")\n",
    "    log.info(\"covariate_plink_df {} : {}\".format(len(covariate_plink_df.columns),covariate_plink_df.columns))         \n",
    "    log.info(\"-----------\")\n",
    "    log.info(\"conditional_omnibus_list {} : {}\".format(len(conditional_omnibus_list),conditional_omnibus_list))\n",
    "    log.info(\"-----------\")\n",
    "    log.info(\"covariate_omnibus_df {} : {}\".format(len(covariate_omnibus_df.columns),covariate_omnibus_df.columns))    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "> ----------------------------Step 1-5 Save covariate and conditional list----------------------------\n"
     ]
    }
   ],
   "source": [
    "    log.info_head(\"Step {}-5 Save covariate and conditional list\".format(idx))\n",
    "        \n",
    "    pd.Series(conditional_plink_list).to_csv(data_out_assoc_phenotype_path+'step_{:02d}.plink.cond'.format(idx),index=None,sep='\\t',header=False)\n",
    "    pd.Series(conditional_omnibus_list).to_csv(data_out_assoc_phenotype_path+'step_{:02d}.omnibus.cond'.format(idx),index=None,sep='\\t',header=False)\n",
    "    \n",
    "    plink_KCHIP_HLA_AA_SNP_1000G_fam.iloc[:,:2].merge(right=covariate_plink_df_drop,left_on='IID',right_index=True).fillna(-9).to_csv(data_out_assoc_phenotype_path+'step_{:02d}.plink.covar'.format(idx),index=None,sep='\\t')\n",
    "    plink_KCHIP_HLA_AA_SNP_1000G_fam.iloc[:,:2].merge(right=covariate_omnibus_df_drop,left_on='IID',right_index=True).fillna(-9).to_csv(data_out_assoc_phenotype_path+'step_{:02d}.omnibus.covar'.format(idx),index=None,sep='\\t')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "            conditional_variant_list_phased_SNP.append(conditional_variant)\n",
    "        log.info(\"calculating r2\")\n",
    "        r2_list=plink_KCHIP_HLA_AA_SNP_1000G_r2_list_dict.get(conditional_variant)\n",
    "        if r2_list is None: \n",
    "            r2_list=[]\n",
    "            for idx_bim,(SNP,row) in enumerate(plink_KCHIP_HLA_AA_SNP_1000G_bim.iterrows()):\n",
    "                r2=pearsonr(plink_KCHIP_HLA_AA_SNP_1000G.get_geno_marker(conditional_variant),plink_KCHIP_HLA_AA_SNP_1000G.get_geno_marker(SNP))[0]**2\n",
    "                r2_list.append(r2)\n",
    "            plink_KCHIP_HLA_AA_SNP_1000G_r2_list_dict[conditional_variant]=r2_list\n",
    "        r2_df=pd.DataFrame(r2_list,index=plink_KCHIP_HLA_AA_SNP_1000G_bim.index)             \n",
    "        plink_KCHIP_HLA_AA_SNP_1000G_HLAassign_count=plink_KCHIP_HLA_AA_SNP_1000G_HLAassign[r2_df[0]>0.7].sum(axis=0).drop('pos')\n",
    "        log.info(\"r2 result:  \\n{}\".format(plink_KCHIP_HLA_AA_SNP_1000G_HLAassign_count))\n",
    "        for HLA_name in plink_KCHIP_HLA_AA_SNP_1000G_HLAassign_count[plink_KCHIP_HLA_AA_SNP_1000G_HLAassign_count>0].index.tolist():\n",
    "            log.info(\"{} added to conditional_variant_list_HLA_AA\".format(HLA_name))\n",
    "            conditional_variant_list_HLA_AA.append(HLA_name)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(59468,\n",
       " 0.05190567456517137,\n",
       " 0.755766030830075,\n",
       " 0.06867955484604434,\n",
       " 0.9452446926391793)"
      ]
     },
     "execution_count": 151,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model_result.nobs,model_result.params.iloc[-1],model_result.bse.iloc[-1],model_result.tvalues.iloc[-1],model_result.pvalues.iloc[-1]#.summary2()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                 Generalized Linear Model Regression Results                  \n",
      "==============================================================================\n",
      "Dep. Variable:                  pheno   No. Observations:                59468\n",
      "Model:                            GLM   Df Residuals:                    59425\n",
      "Model Family:                Gaussian   Df Model:                           42\n",
      "Link Function:               identity   Scale:                          5288.2\n",
      "Method:                          IRLS   Log-Likelihood:            -3.3928e+05\n",
      "Date:                Sun, 01 Mar 2020   Deviance:                   3.1425e+08\n",
      "Time:                        22:46:35   Pearson chi2:                 3.14e+08\n",
      "No. Iterations:                     3                                         \n",
      "Covariance Type:            nonrobust                                         \n",
      "==============================================================================\n",
      "                 coef    std err          z      P>|z|      [0.025      0.975]\n",
      "------------------------------------------------------------------------------\n",
      "const         16.0973      4.012      4.013      0.000       8.235      23.960\n",
      "x1            11.6024     18.165      0.639      0.523     -24.001      47.205\n",
      "x2           -63.6028     18.655     -3.409      0.001    -100.166     -27.040\n",
      "x3             5.6825     19.193      0.296      0.767     -31.934      43.299\n",
      "x4            18.2657     19.317      0.946      0.344     -19.596      56.127\n",
      "x5             1.9840      0.037     54.058      0.000       1.912       2.056\n",
      "x6            -7.4229      0.624    -11.896      0.000      -8.646      -6.200\n",
      "x7          1.609e-12   5.21e-13      3.088      0.002    5.87e-13    2.63e-12\n",
      "x8           1.04e-13   1.55e-13      0.672      0.502   -1.99e-13    4.07e-13\n",
      "x9           -13.7414     25.489     -0.539      0.590     -63.699      36.216\n",
      "x10          -53.0547     36.367     -1.459      0.145    -124.332      18.222\n",
      "x11          -16.9566     16.402     -1.034      0.301     -49.105      15.192\n",
      "x12          -59.5173     41.495     -1.434      0.151    -140.846      21.811\n",
      "x13          -26.5920     18.190     -1.462      0.144     -62.244       9.060\n",
      "x14            6.7096      5.908      1.136      0.256      -4.870      18.289\n",
      "x15            1.9639      4.429      0.443      0.657      -6.716      10.644\n",
      "x16            2.5330      6.671      0.380      0.704     -10.541      15.607\n",
      "x17           15.8219      8.271      1.913      0.056      -0.390      32.034\n",
      "x18           -0.2998      8.372     -0.036      0.971     -16.708      16.108\n",
      "x19            9.0341     25.746      0.351      0.726     -41.427      59.495\n",
      "x20            8.2913      5.598      1.481      0.139      -2.682      19.264\n",
      "x21           -0.2735      3.593     -0.076      0.939      -7.315       6.768\n",
      "x22          -17.6143     17.327     -1.017      0.309     -51.574      16.346\n",
      "x23           -0.2004      6.616     -0.030      0.976     -13.167      12.766\n",
      "x24            2.4811     13.998      0.177      0.859     -24.955      29.917\n",
      "x25            3.8326     10.075      0.380      0.704     -15.914      23.579\n",
      "x26           -1.6874      6.034     -0.280      0.780     -13.514      10.139\n",
      "x27           10.3293     10.739      0.962      0.336     -10.719      31.378\n",
      "x28            8.0178      5.291      1.515      0.130      -2.353      18.389\n",
      "x29           -3.3305      6.165     -0.540      0.589     -15.415       8.754\n",
      "x30           14.7435      8.316      1.773      0.076      -1.556      31.043\n",
      "x31            6.9540      6.433      1.081      0.280      -5.655      19.563\n",
      "x32          -10.8942     11.842     -0.920      0.358     -34.104      12.316\n",
      "x33            4.3415     13.701      0.317      0.751     -22.511      31.194\n",
      "x34            0.7247      8.988      0.081      0.936     -16.892      18.341\n",
      "x35            0.7247      8.988      0.081      0.936     -16.892      18.341\n",
      "x36          -10.8942     11.842     -0.920      0.358     -34.104      12.316\n",
      "x37            6.9540      6.433      1.081      0.280      -5.655      19.563\n",
      "x38          -30.1680     15.419     -1.957      0.050     -60.389       0.053\n",
      "x39           -3.3305      6.165     -0.540      0.589     -15.415       8.754\n",
      "x40          -11.5515     11.508     -1.004      0.315     -34.108      11.004\n",
      "x41            9.2068      6.861      1.342      0.180      -4.240      22.654\n",
      "x42           26.0207     24.103      1.080      0.280     -21.220      73.261\n",
      "x43            4.1303      5.734      0.720      0.471      -7.109      15.369\n",
      "x44            1.7723      5.200      0.341      0.733      -8.420      11.965\n",
      "x45           15.5591     10.257      1.517      0.129      -4.545      35.663\n",
      "x46           -9.6565      5.665     -1.705      0.088     -20.760       1.447\n",
      "x47            2.0458      3.961      0.516      0.606      -5.719       9.810\n",
      "x48            2.0458      3.961      0.516      0.606      -5.719       9.810\n",
      "x49            2.0458      3.961      0.516      0.606      -5.719       9.810\n",
      "x50            2.0458      3.961      0.516      0.606      -5.719       9.810\n",
      "x51            2.0458      3.961      0.516      0.606      -5.719       9.810\n",
      "x52          -10.9385     18.059     -0.606      0.545     -46.334      24.457\n",
      "x53          -36.9127     22.985     -1.606      0.108     -81.962       8.137\n",
      "x54           22.4603      8.301      2.706      0.007       6.191      38.730\n",
      "x55          -12.5948     26.156     -0.482      0.630     -63.860      38.671\n",
      "x56          -17.5556     14.899     -1.178      0.239     -46.757      11.646\n",
      "x57          -26.5920     18.190     -1.462      0.144     -62.244       9.060\n",
      "x58          -32.9180     28.470     -1.156      0.248     -88.718      22.882\n",
      "x59            9.0341     25.746      0.351      0.726     -41.427      59.495\n",
      "x60          -11.1389     13.758     -0.810      0.418     -38.104      15.826\n",
      "x61           18.4656     14.297      1.292      0.196      -9.555      46.487\n",
      "x62           -0.2004      6.616     -0.030      0.976     -13.167      12.766\n",
      "x63            8.0178      5.291      1.515      0.130      -2.353      18.389\n",
      "x64           17.6055     12.440      1.415      0.157      -6.777      41.988\n",
      "x65           21.0238     11.485      1.831      0.067      -1.487      43.534\n",
      "x66           22.8073     11.675      1.954      0.051      -0.075      45.690\n",
      "x67            8.0178      5.291      1.515      0.130      -2.353      18.389\n",
      "x68           15.8219      8.271      1.913      0.056      -0.390      32.034\n",
      "x69            4.5275      4.447      1.018      0.309      -4.188      13.243\n",
      "x70           15.9649      4.248      3.758      0.000       7.639      24.291\n",
      "x71           14.6347      9.324      1.570      0.117      -3.640      32.909\n",
      "x72           15.9649      4.248      3.758      0.000       7.639      24.291\n",
      "x73           -8.3575      9.550     -0.875      0.382     -27.075      10.360\n",
      "x74           -3.3305      6.165     -0.540      0.589     -15.415       8.754\n",
      "x75           -4.7370     11.378     -0.416      0.677     -27.037      17.563\n",
      "x76           -8.4916     14.662     -0.579      0.562     -37.228      20.245\n",
      "x77          -12.6939      9.991     -1.271      0.204     -32.275       6.887\n",
      "x78            8.2913      5.598      1.481      0.139      -2.682      19.264\n",
      "x79           -1.7841      7.707     -0.231      0.817     -16.889      13.321\n",
      "x80           -3.3305      6.165     -0.540      0.589     -15.415       8.754\n",
      "x81          -15.1263     11.407     -1.326      0.185     -37.484       7.232\n",
      "x82           -3.3305      6.165     -0.540      0.589     -15.415       8.754\n",
      "x83            9.3360     12.536      0.745      0.456     -15.234      33.906\n",
      "x84           29.5717     43.977      0.672      0.501     -56.622     115.766\n",
      "x85            2.7390      6.542      0.419      0.675     -10.084      15.562\n",
      "x86            2.7390      6.542      0.419      0.675     -10.084      15.562\n",
      "x87           13.4694      6.200      2.172      0.030       1.317      25.621\n",
      "x88            9.5185      3.558      2.676      0.007       2.546      16.491\n",
      "x89            9.5185      3.558      2.676      0.007       2.546      16.491\n",
      "x90            9.5185      3.558      2.676      0.007       2.546      16.491\n",
      "x91            6.7113      3.753      1.788      0.074      -0.645      14.068\n",
      "x92            9.5185      3.558      2.676      0.007       2.546      16.491\n",
      "x93           -2.1838      4.989     -0.438      0.662     -11.962       7.595\n",
      "x94           11.7023      6.318      1.852      0.064      -0.681      24.086\n",
      "x95            9.5185      3.558      2.676      0.007       2.546      16.491\n",
      "x96           -0.2004      6.616     -0.030      0.976     -13.167      12.766\n",
      "x97           -1.8878      7.084     -0.266      0.790     -15.772      11.996\n",
      "x98            6.7113      3.753      1.788      0.074      -0.645      14.068\n",
      "x99           -1.8878      7.084     -0.266      0.790     -15.772      11.996\n",
      "x100          11.7023      6.318      1.852      0.064      -0.681      24.086\n",
      "x101          -2.1838      4.989     -0.438      0.662     -11.962       7.595\n",
      "x102           6.7113      3.753      1.788      0.074      -0.645      14.068\n",
      "x103          -4.0716      8.779     -0.464      0.643     -21.278      13.135\n",
      "x104         -14.7851      0.679    -21.783      0.000     -16.115     -13.455\n",
      "x105          -0.2413      0.735     -0.328      0.743      -1.681       1.198\n",
      "==============================================================================\n",
      "                 Generalized Linear Model Regression Results                  \n",
      "==============================================================================\n",
      "Dep. Variable:                  pheno   No. Observations:                59468\n",
      "Model:                            GLM   Df Residuals:                    59425\n",
      "Model Family:                Gaussian   Df Model:                           42\n",
      "Link Function:               identity   Scale:                          5288.2\n",
      "Method:                          IRLS   Log-Likelihood:            -3.3928e+05\n",
      "Date:                Sun, 01 Mar 2020   Deviance:                   3.1425e+08\n",
      "Time:                        22:46:37   Pearson chi2:                 3.14e+08\n",
      "No. Iterations:                     3                                         \n",
      "Covariance Type:            nonrobust                                         \n",
      "==============================================================================\n",
      "                 coef    std err          z      P>|z|      [0.025      0.975]\n",
      "------------------------------------------------------------------------------\n",
      "const         16.0975      4.012      4.013      0.000       8.235      23.960\n",
      "x1            11.6024     18.165      0.639      0.523     -24.000      47.205\n",
      "x2           -63.6031     18.655     -3.409      0.001    -100.166     -27.040\n",
      "x3             5.6828     19.193      0.296      0.767     -31.934      43.300\n",
      "x4            18.2661     19.317      0.946      0.344     -19.595      56.127\n",
      "x5             1.9840      0.037     54.058      0.000       1.912       2.056\n",
      "x6            -7.4229      0.624    -11.896      0.000      -8.646      -6.200\n",
      "x7         -1.192e-12   4.32e-13     -2.756      0.006   -2.04e-12   -3.44e-13\n",
      "x8         -4.189e-14   3.14e-13     -0.133      0.894   -6.58e-13    5.74e-13\n",
      "x9           -13.7409     25.489     -0.539      0.590     -63.698      36.217\n",
      "x10          -53.0535     36.367     -1.459      0.145    -124.331      18.224\n",
      "x11          -16.9566     16.402     -1.034      0.301     -49.105      15.192\n",
      "x12          -59.5161     41.495     -1.434      0.151    -140.844      21.812\n",
      "x13          -26.5917     18.190     -1.462      0.144     -62.244       9.060\n",
      "x14            6.7098      5.908      1.136      0.256      -4.870      18.289\n",
      "x15            1.9635      4.429      0.443      0.657      -6.716      10.643\n",
      "x16            2.5327      6.671      0.380      0.704     -10.541      15.607\n",
      "x17           15.8217      8.271      1.913      0.056      -0.390      32.033\n",
      "x18           -0.2986      8.372     -0.036      0.972     -16.707      16.109\n",
      "x19            9.0330     25.746      0.351      0.726     -41.428      59.494\n",
      "x20            8.2914      5.598      1.481      0.139      -2.681      19.264\n",
      "x21           -0.2737      3.593     -0.076      0.939      -7.315       6.768\n",
      "x22          -17.6140     17.327     -1.017      0.309     -51.574      16.346\n",
      "x23           -0.2006      6.616     -0.030      0.976     -13.167      12.766\n",
      "x24            2.4805     13.998      0.177      0.859     -24.956      29.917\n",
      "x25            3.8323     10.075      0.380      0.704     -15.914      23.579\n",
      "x26           -1.6872      6.034     -0.280      0.780     -13.514      10.139\n",
      "x27           10.3293     10.739      0.962      0.336     -10.719      31.378\n",
      "x28            8.0177      5.291      1.515      0.130      -2.353      18.389\n",
      "x29           -3.3299      6.165     -0.540      0.589     -15.414       8.754\n",
      "x30           14.7439      8.316      1.773      0.076      -1.556      31.044\n",
      "x31            6.9535      6.433      1.081      0.280      -5.655      19.562\n",
      "x32          -10.8939     11.842     -0.920      0.358     -34.104      12.316\n",
      "x33            4.3414     13.701      0.317      0.751     -22.511      31.194\n",
      "x34            0.7250      8.988      0.081      0.936     -16.891      18.341\n",
      "x35            0.7250      8.988      0.081      0.936     -16.891      18.341\n",
      "x36          -10.8939     11.842     -0.920      0.358     -34.104      12.316\n",
      "x37            6.9535      6.433      1.081      0.280      -5.655      19.562\n",
      "x38          -30.1677     15.419     -1.957      0.050     -60.389       0.053\n",
      "x39           -3.3299      6.165     -0.540      0.589     -15.414       8.754\n",
      "x40          -11.5513     11.508     -1.004      0.316     -34.107      11.005\n",
      "x41            9.2067      6.861      1.342      0.180      -4.240      22.654\n",
      "x42           26.0210     24.103      1.080      0.280     -21.220      73.262\n",
      "x43            4.1307      5.734      0.720      0.471      -7.108      15.370\n",
      "x44            1.7721      5.200      0.341      0.733      -8.420      11.965\n",
      "x45           15.5599     10.257      1.517      0.129      -4.544      35.664\n",
      "x46           -9.6571      5.665     -1.705      0.088     -20.761       1.446\n",
      "x47            2.0458      3.961      0.516      0.606      -5.718       9.810\n",
      "x48            2.0458      3.961      0.516      0.606      -5.718       9.810\n",
      "x49            2.0458      3.961      0.516      0.606      -5.718       9.810\n",
      "x50            2.0458      3.961      0.516      0.606      -5.718       9.810\n",
      "x51            2.0458      3.961      0.516      0.606      -5.718       9.810\n",
      "x52          -10.9369     18.059     -0.606      0.545     -46.332      24.459\n",
      "x53          -36.9124     22.985     -1.606      0.108     -81.962       8.137\n",
      "x54           22.4603      8.301      2.706      0.007       6.191      38.730\n",
      "x55          -12.5951     26.156     -0.482      0.630     -63.861      38.671\n",
      "x56          -17.5556     14.899     -1.178      0.239     -46.757      11.646\n",
      "x57          -26.5917     18.190     -1.462      0.144     -62.244       9.060\n",
      "x58          -32.9176     28.470     -1.156      0.248     -88.717      22.882\n",
      "x59            9.0330     25.746      0.351      0.726     -41.428      59.494\n",
      "x60          -11.1376     13.758     -0.810      0.418     -38.102      15.827\n",
      "x61           18.4655     14.297      1.292      0.196      -9.555      46.486\n",
      "x62           -0.2006      6.616     -0.030      0.976     -13.167      12.766\n",
      "x63            8.0177      5.291      1.515      0.130      -2.353      18.389\n",
      "x64           17.6055     12.440      1.415      0.157      -6.777      41.988\n",
      "x65           21.0239     11.485      1.831      0.067      -1.486      43.534\n",
      "x66           22.8077     11.675      1.954      0.051      -0.075      45.690\n",
      "x67            8.0177      5.291      1.515      0.130      -2.353      18.389\n",
      "x68           15.8217      8.271      1.913      0.056      -0.390      32.033\n",
      "x69            4.5274      4.447      1.018      0.309      -4.188      13.243\n",
      "x70           15.9647      4.248      3.758      0.000       7.639      24.291\n",
      "x71           14.6341      9.324      1.570      0.117      -3.640      32.909\n",
      "x72           15.9647      4.248      3.758      0.000       7.639      24.291\n",
      "x73           -8.3572      9.550     -0.875      0.382     -27.075      10.360\n",
      "x74           -3.3299      6.165     -0.540      0.589     -15.414       8.754\n",
      "x75           -4.7372     11.378     -0.416      0.677     -27.037      17.563\n",
      "x76           -8.4906     14.662     -0.579      0.563     -37.227      20.246\n",
      "x77          -12.6932      9.991     -1.271      0.204     -32.274       6.888\n",
      "x78            8.2914      5.598      1.481      0.139      -2.681      19.264\n",
      "x79           -1.7836      7.707     -0.231      0.817     -16.889      13.322\n",
      "x80           -3.3299      6.165     -0.540      0.589     -15.414       8.754\n",
      "x81          -15.1266     11.407     -1.326      0.185     -37.484       7.231\n",
      "x82           -3.3299      6.165     -0.540      0.589     -15.414       8.754\n",
      "x83            9.3345     12.536      0.745      0.456     -15.235      33.904\n",
      "x84           29.5667     43.977      0.672      0.501     -56.627     115.761\n",
      "x85            2.7371      6.542      0.418      0.676     -10.086      15.560\n",
      "x86            2.7371      6.542      0.418      0.676     -10.086      15.560\n",
      "x87           13.4695      6.200      2.172      0.030       1.317      25.621\n",
      "x88            9.5188      3.558      2.676      0.007       2.546      16.491\n",
      "x89            9.5188      3.558      2.676      0.007       2.546      16.491\n",
      "x90            9.5188      3.558      2.676      0.007       2.546      16.491\n",
      "x91            6.7116      3.753      1.788      0.074      -0.645      14.068\n",
      "x92            9.5188      3.558      2.676      0.007       2.546      16.491\n",
      "x93           -2.1841      4.989     -0.438      0.662     -11.962       7.594\n",
      "x94           11.7029      6.318      1.852      0.064      -0.681      24.086\n",
      "x95            9.5188      3.558      2.676      0.007       2.546      16.491\n",
      "x96           -0.2006      6.616     -0.030      0.976     -13.167      12.766\n",
      "x97           -1.8878      7.084     -0.266      0.790     -15.772      11.996\n",
      "x98            6.7116      3.753      1.788      0.074      -0.645      14.068\n",
      "x99           -1.8878      7.084     -0.266      0.790     -15.772      11.996\n",
      "x100          11.7029      6.318      1.852      0.064      -0.681      24.086\n",
      "x101          -2.1841      4.989     -0.438      0.662     -11.962       7.594\n",
      "x102           6.7116      3.753      1.788      0.074      -0.645      14.068\n",
      "x103          -4.0719      8.779     -0.464      0.643     -21.279      13.135\n",
      "x104         -14.7851      0.679    -21.783      0.000     -16.115     -13.455\n",
      "x105          -0.2467      0.735     -0.336      0.737      -1.686       1.193\n",
      "==============================================================================\n",
      "                 Generalized Linear Model Regression Results                  \n",
      "==============================================================================\n",
      "Dep. Variable:                  pheno   No. Observations:                59468\n",
      "Model:                            GLM   Df Residuals:                    59425\n",
      "Model Family:                Gaussian   Df Model:                           42\n",
      "Link Function:               identity   Scale:                          5288.2\n",
      "Method:                          IRLS   Log-Likelihood:            -3.3928e+05\n",
      "Date:                Sun, 01 Mar 2020   Deviance:                   3.1425e+08\n",
      "Time:                        22:46:38   Pearson chi2:                 3.14e+08\n",
      "No. Iterations:                     3                                         \n",
      "Covariance Type:            nonrobust                                         \n",
      "==============================================================================\n",
      "                 coef    std err          z      P>|z|      [0.025      0.975]\n",
      "------------------------------------------------------------------------------\n",
      "const         16.0884      4.012      4.011      0.000       8.226      23.951\n",
      "x1            11.5919     18.165      0.638      0.523     -24.011      47.195\n",
      "x2           -63.5722     18.655     -3.408      0.001    -100.135     -27.009\n",
      "x3             5.6695     19.193      0.295      0.768     -31.947      43.286\n",
      "x4            18.2535     19.317      0.945      0.345     -19.608      56.115\n",
      "x5             1.9840      0.037     54.058      0.000       1.912       2.056\n",
      "x6            -7.4238      0.624    -11.897      0.000      -8.647      -6.201\n",
      "x7          4.407e-13   9.03e-13      0.488      0.625   -1.33e-12    2.21e-12\n",
      "x8          1.078e-12   5.43e-13      1.986      0.047    1.44e-14    2.14e-12\n",
      "x9           -13.7703     25.489     -0.540      0.589     -63.728      36.187\n",
      "x10          -53.1228     36.367     -1.461      0.144    -124.400      18.154\n",
      "x11          -16.9553     16.402     -1.034      0.301     -49.104      15.193\n",
      "x12          -59.5843     41.495     -1.436      0.151    -140.913      21.744\n",
      "x13          -26.6098     18.190     -1.463      0.144     -62.262       9.042\n",
      "x14            6.7018      5.908      1.134      0.257      -4.878      18.281\n",
      "x15            1.9822      4.429      0.448      0.654      -6.698      10.662\n",
      "x16            2.5483      6.671      0.382      0.702     -10.526      15.623\n",
      "x17           15.8343      8.272      1.914      0.056      -0.378      32.046\n",
      "x18           -0.3618      8.372     -0.043      0.966     -16.770      16.047\n",
      "x19            9.0936     25.746      0.353      0.724     -41.367      59.554\n",
      "x20            8.2876      5.598      1.480      0.139      -2.685      19.260\n",
      "x21           -0.2611      3.593     -0.073      0.942      -7.303       6.780\n",
      "x22          -17.6303     17.327     -1.018      0.309     -51.590      16.330\n",
      "x23           -0.1885      6.616     -0.028      0.977     -13.155      12.778\n",
      "x24            2.5131     13.998      0.180      0.858     -24.923      29.949\n",
      "x25            3.8505     10.075      0.382      0.702     -15.896      23.597\n",
      "x26           -1.6980      6.034     -0.281      0.778     -13.525      10.129\n",
      "x27           10.3322     10.739      0.962      0.336     -10.716      31.381\n",
      "x28            8.0265      5.291      1.517      0.129      -2.345      18.398\n",
      "x29           -3.3610      6.166     -0.545      0.586     -15.445       8.723\n",
      "x30           14.7245      8.316      1.771      0.077      -1.575      31.024\n",
      "x31            6.9784      6.433      1.085      0.278      -5.630      19.587\n",
      "x32          -10.9091     11.842     -0.921      0.357     -34.119      12.301\n",
      "x33            4.3432     13.701      0.317      0.751     -22.509      31.196\n",
      "x34            0.7094      8.988      0.079      0.937     -16.907      18.326\n",
      "x35            0.7094      8.988      0.079      0.937     -16.907      18.326\n",
      "x36          -10.9091     11.842     -0.921      0.357     -34.119      12.301\n",
      "x37            6.9784      6.433      1.085      0.278      -5.630      19.587\n",
      "x38          -30.1848     15.419     -1.958      0.050     -60.406       0.036\n",
      "x39           -3.3610      6.166     -0.545      0.586     -15.445       8.723\n",
      "x40          -11.5654     11.508     -1.005      0.315     -34.121      10.991\n",
      "x41            9.2111      6.861      1.343      0.179      -4.236      22.658\n",
      "x42           26.0058     24.103      1.079      0.281     -21.235      73.246\n",
      "x43            4.1125      5.734      0.717      0.473      -7.126      15.351\n",
      "x44            1.7806      5.200      0.342      0.732      -8.412      11.973\n",
      "x45           15.5172     10.257      1.513      0.130      -4.587      35.621\n",
      "x46           -9.6241      5.665     -1.699      0.089     -20.727       1.479\n",
      "x47            2.0417      3.961      0.515      0.606      -5.723       9.806\n",
      "x48            2.0417      3.961      0.515      0.606      -5.723       9.806\n",
      "x49            2.0417      3.961      0.515      0.606      -5.723       9.806\n",
      "x50            2.0417      3.961      0.515      0.606      -5.723       9.806\n",
      "x51            2.0417      3.961      0.515      0.606      -5.723       9.806\n",
      "x52          -11.0231     18.059     -0.610      0.542     -46.419      24.373\n",
      "x53          -36.9324     22.985     -1.607      0.108     -81.982       8.118\n",
      "x54           22.4577      8.301      2.705      0.007       6.188      38.727\n",
      "x55          -12.5756     26.156     -0.481      0.631     -63.841      38.690\n",
      "x56          -17.5553     14.899     -1.178      0.239     -46.757      11.646\n",
      "x57          -26.6098     18.190     -1.463      0.144     -62.262       9.042\n",
      "x58          -32.9430     28.470     -1.157      0.247     -88.743      22.857\n",
      "x59            9.0936     25.746      0.353      0.724     -41.367      59.554\n",
      "x60          -11.2116     13.758     -0.815      0.415     -38.177      15.754\n",
      "x61           18.4683     14.297      1.292      0.196      -9.553      46.489\n",
      "x62           -0.1885      6.616     -0.028      0.977     -13.155      12.778\n",
      "x63            8.0265      5.291      1.517      0.129      -2.345      18.398\n",
      "x64           17.6079     12.440      1.415      0.157      -6.774      41.990\n",
      "x65           21.0173     11.485      1.830      0.067      -1.493      43.528\n",
      "x66           22.7908     11.675      1.952      0.051      -0.091      45.673\n",
      "x67            8.0265      5.291      1.517      0.129      -2.345      18.398\n",
      "x68           15.8343      8.272      1.914      0.056      -0.378      32.046\n",
      "x69            4.5324      4.447      1.019      0.308      -4.183      13.248\n",
      "x70           15.9787      4.248      3.761      0.000       7.652      24.305\n",
      "x71           14.6652      9.324      1.573      0.116      -3.609      32.940\n",
      "x72           15.9787      4.248      3.761      0.000       7.652      24.305\n",
      "x73           -8.3700      9.550     -0.876      0.381     -27.088      10.348\n",
      "x74           -3.3610      6.166     -0.545      0.586     -15.445       8.723\n",
      "x75           -4.7233     11.378     -0.415      0.678     -27.023      17.577\n",
      "x76           -8.5439     14.662     -0.583      0.560     -37.280      20.192\n",
      "x77          -12.7327      9.991     -1.274      0.203     -32.314       6.849\n",
      "x78            8.2876      5.598      1.480      0.139      -2.685      19.260\n",
      "x79           -1.8079      7.707     -0.235      0.815     -16.913      13.297\n",
      "x80           -3.3610      6.166     -0.545      0.586     -15.445       8.723\n",
      "x81          -15.1102     11.407     -1.325      0.185     -37.468       7.248\n",
      "x82           -3.3610      6.166     -0.545      0.586     -15.445       8.723\n",
      "x83            9.4156     12.536      0.751      0.453     -15.154      33.985\n",
      "x84           29.8408     43.978      0.679      0.497     -56.354     116.035\n",
      "x85            2.8450      6.543      0.435      0.664      -9.979      15.669\n",
      "x86            2.8450      6.543      0.435      0.664      -9.979      15.669\n",
      "x87           13.4663      6.200      2.172      0.030       1.314      25.618\n",
      "x88            9.4995      3.558      2.670      0.008       2.527      16.472\n",
      "x89            9.4995      3.558      2.670      0.008       2.527      16.472\n",
      "x90            9.4995      3.558      2.670      0.008       2.527      16.472\n",
      "x91            6.6987      3.753      1.785      0.074      -0.658      14.055\n",
      "x92            9.4995      3.558      2.670      0.008       2.527      16.472\n",
      "x93           -2.1663      4.989     -0.434      0.664     -11.945       7.612\n",
      "x94           11.6658      6.318      1.846      0.065      -0.718      24.049\n",
      "x95            9.4995      3.558      2.670      0.008       2.527      16.472\n",
      "x96           -0.1885      6.616     -0.028      0.977     -13.155      12.778\n",
      "x97           -1.8865      7.084     -0.266      0.790     -15.771      11.998\n",
      "x98            6.6987      3.753      1.785      0.074      -0.658      14.055\n",
      "x99           -1.8865      7.084     -0.266      0.790     -15.771      11.998\n",
      "x100          11.6658      6.318      1.846      0.065      -0.718      24.049\n",
      "x101          -2.1663      4.989     -0.434      0.664     -11.945       7.612\n",
      "x102           6.6987      3.753      1.785      0.074      -0.658      14.055\n",
      "x103          -4.0528      8.779     -0.462      0.644     -21.259      13.154\n",
      "x104         -14.7823      0.679    -21.779      0.000     -16.113     -13.452\n",
      "x105           0.0519      0.756      0.069      0.945      -1.429       1.533\n",
      "==============================================================================\n"
     ]
    }
   ],
   "source": [
    "for idx_bim,(SNP,row) in enumerate(plink_KCHIP_HLA_AA_SNP_1000G_bim.iterrows()):\n",
    "    y_data=pheno['pheno'].replace(-9,np.nan)\n",
    "    \n",
    "    \n",
    "    x_data_intercept=np.array([np.ones(plink_KCHIP_HLA_AA_SNP_1000G_fam.shape[0])]).transpose()    \n",
    "    x_data_dosage=plink_KCHIP_HLA_AA_SNP_1000G.get_geno_marker(SNP).astype(float);x_data_dosage[x_data_dosage==-1]=np.nan\n",
    "    x_data_dosage=np.expand_dims(x_data_dosage,axis=1)\n",
    "    x_data_covariate=covariate_plink_df.values\n",
    "    #print(x_data_intercept.shape,covariate_plink_df.values.shape)\n",
    "    x_data=np.concatenate([x_data_intercept,x_data_covariate,x_data_dosage],axis=1)#[~x_y_data_nan]\n",
    "    x_data_null=np.concatenate([x_data_intercept,x_data_covariate],axis=1)#[~x_y_data_nan]\n",
    "    \n",
    "    family=(sm.families.Gaussian() if phenotype_type=='continuous' else sm.families.Binomial())\n",
    "    model=sm.GLM(y_data,x_data, family=family,missing='drop')\n",
    "    model_null=sm.GLM(y_data,x_data_null, family=family,missing='drop')\n",
    "    model_result=model.fit()\n",
    "    model_result_null=model_null.fit()\n",
    "    \n",
    "    print(model_result.summary())\n",
    "    if idx_bim==2:\n",
    "        break\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array([1., 0., 0., ..., 0., 0., 0.]), array([[1.],\n",
       "        [0.],\n",
       "        [0.],\n",
       "        ...,\n",
       "        [0.],\n",
       "        [0.],\n",
       "        [0.]]))"
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x_data_dosage,np.expand_dims(x_data_dosage,axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "    x_data_null=np.concatenate([x_data_intercept,x_data_covar,x_data_condition,x_data_aa_marker],axis=1)[~x_y_data_nan]\n",
    "    x_data_alt=np.concatenate([x_data_intercept,x_data_covar,x_data_condition],axis=1)[~x_y_data_nan]\n",
    "    y_data=y_data[~x_y_data_nan]\n",
    "    \n",
    "    \n",
    "    family=(sm.families.Gaussian() if assoc=='linear' else sm.families.Binomial())\n",
    "    model_null = sm.GLM(y_data,x_data_null, family=family,missing='raise')\n",
    "    model_alt = sm.GLM(y_data,x_data_alt, family=family,missing='raise')\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "> -------------------------------------Step 1-6 Plink Association-------------------------------------\n",
      "> plink2 --bfile data/genotype/4_merge/KCHIP_HLA_SNP_1000G --linear --pheno data/out_pheno/ALP.phe --covar data/out_assoc/ALP/step_01.plink.covar --condition-list data/out_assoc/ALP/step_01.plink.cond --out data/out_assoc/ALP/step_01.plink --covar-variance-standardize --threads 40\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "------RUN-----\n",
      "plink2 --bfile data/genotype/4_merge/KCHIP_HLA_SNP_1000G --linear --pheno data/out_pheno/ALP.phe --covar data/out_assoc/ALP/step_01.plink.covar --condition-list data/out_assoc/ALP/step_01.plink.cond --out data/out_assoc/ALP/step_01.plink --covar-variance-standardize --threads 40\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "> PLINK v2.00a2.3LM 64-bit Intel (24 Jan 2020)   www.cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to data/out_assoc/ALP/step_01.plink.log.\n",
      "Options in effect:\n",
      "  --bfile data/genotype/4_merge/KCHIP_HLA_SNP_1000G\n",
      "  --condition-list data/out_assoc/ALP/step_01.plink.cond\n",
      "  --covar data/out_assoc/ALP/step_01.plink.covar\n",
      "  --covar-variance-standardize\n",
      "  --glm\n",
      "  --out data/out_assoc/ALP/step_01.plink\n",
      "  --pheno data/out_pheno/ALP.phe\n",
      "  --threads 40\n",
      "\n",
      "Start time: Sun Mar  1 22:22:24 2020\n",
      "257299 MiB RAM detected; reserving 128649 MiB for main workspace.\n",
      "Using up to 40 threads (change this with --threads).\n",
      "125673 samples (80026 females, 45647 males; 125673 founders) loaded from\n",
      "data/genotype/4_merge/KCHIP_HLA_SNP_1000G.fam.\n",
      "74320 variants loaded from data/genotype/4_merge/KCHIP_HLA_SNP_1000G.bim.\n",
      "2 phenotypes loaded (1 binary, 1 quantitative).\n",
      "40 covariates loaded from data/out_assoc/ALP/step_01.plink.covar.\n",
      "--covar-variance-standardize: 40 covariates transformed.\n",
      "Calculating allele frequencies... 88done.\n",
      "--condition-list: 1 variant ID loaded.\n",
      "--glm: Skipping case/control phenotype 'PHENO1' since all samples are controls.\n",
      "End time: Sun Mar  1 22:22:25 2020\n",
      "\n",
      "> Warning: Not including covariate 'AS' in --glm regression on phenotype\n",
      "'PHENO2'.\n",
      "Warning: Not including covariate 'NC' in --glm regression on phenotype\n",
      "'PHENO2'.\n",
      "Warning: Skipping --glm regression on phenotype 'PHENO2', since covariate\n",
      "correlation matrix could not be inverted (VIF_INFINITE). You may want to remove\n",
      "redundant covariates and try again.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "    log.info_head(\"Step {}-6 Plink Association\".format(idx))\n",
    "\n",
    "\n",
    "    command='plink2 --bfile {bfile} {assoc_mode} --pheno {pheno} {covar_on} {covar} --condition-list {condition_list} --out {out} --covar-variance-standardize --threads 40'.format(bfile=plink_KCHIP_HLA_SNP_1000G_path,\n",
    "                                                                                                                        assoc_mode='--logistic' if phenotype_type=='binary' else '--linear',\n",
    "                                                                                                                         pheno=data_out_pheno_path+phenotype_name+'.phe',\n",
    "                                                                                                                         covar_on='--covar' if len(covariate_plink_df.columns)!=0 else '',\n",
    "                                                                                                                         covar=data_out_assoc_phenotype_path+'step_{:02d}.plink.covar'.format(idx) if len(covariate_plink_df.columns)!=0 else '',\n",
    "                                                                                                                         condition_list=data_out_assoc_phenotype_path+'step_{:02d}.plink.cond'.format(idx),\n",
    "                                                                                                                         out=data_out_assoc_phenotype_path+'step_{:02d}.plink'.format(idx)                                                                         \n",
    "                                                                                                                        )\n",
    "    log.info(command)\n",
    "    stdout,stderr=run_subprocess(command,dry=False,quiet=True)\n",
    "    log.info(stdout)\n",
    "    log.error(stderr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "    log.info_head(\"Step {}-6 Plink Association\".format(idx))\n",
    "\n",
    "\n",
    "    command='plink2 --bfile {bfile} {assoc_mode} --pheno {pheno} {covar_on} {covar} --condition-list {condition_list} --out {out} --covar-variance-standardize --threads 40'.format(bfile=plink_KCHIP_HLA_SNP_1000G_path,\n",
    "                                                                                                                        assoc_mode='--logistic' if phenotype_type=='binary' else '--linear',\n",
    "                                                                                                                         pheno=data_out_pheno_path+phenotype_name+'.phe',\n",
    "                                                                                                                         covar_on='--covar' if len(covariate_plink_df.columns)!=0 else '',\n",
    "                                                                                                                         covar=data_out_assoc_phenotype_path+'step_{:02d}.plink.covar'.format(idx) if len(covariate_plink_df.columns)!=0 else '',\n",
    "                                                                                                                         condition_list=data_out_assoc_phenotype_path+'step_{:02d}.plink.cond'.format(idx),\n",
    "                                                                                                                         out=data_out_assoc_phenotype_path+'step_{:02d}.plink'.format(idx)                                                                         \n",
    "                                                                                                                        )\n",
    "    log.info(command)\n",
    "    stdout,stderr=run_subprocess(command,dry=False,quiet=True)\n",
    "    log.info(stdout)\n",
    "    log.error(stderr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "plink_KCHIP_HLA_AA_SNP_1000G_fam.iloc[:,:2].merge(right=covariate_plink_df,left_on='IID',right_index=True).fillna(-9).to_csv(data_out_assoc_phenotype_path+'step_{:02d}.plink.covar'.format(idx),index=None,sep='\\t')\n",
    "plink_KCHIP_HLA_AA_SNP_1000G_fam.iloc[:,:2].merge(right=covariate_omnibus_df,left_on='IID',right_index=True).fillna(-9).to_csv(data_out_assoc_phenotype_path+'step_{:02d}.omnibus.covar'.format(idx),index=None,sep='\\t')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "    covariate_plink_df=covariate_plink_df.reset_index()\n",
    "    covariate_plink=covariate_plink.drop(columns=covariate_plink.iloc[:,1:].columns[get_dupbool(covariate_plink.iloc[:,1:])])\n",
    "    if covariate_plink.shape[1]>1:\n",
    "        covariate_plink[[covariate_plink.columns[0],covariate_plink.columns[0]]+covariate_plink.columns[1:].values.tolist()].fillna(-9).to_csv(data_out_assoc_phenotype_path+'step_{:02d}.plink.covar'.format(idx),index=None,sep='\\t')\n",
    "        #cova"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "    log.info_head(\"Step {}-6 Save covariate and conditional list\".format(idx))\n",
    "        \n",
    "    pd.Series(conditional_plink_list).to_csv(data_out_assoc_phenotype_path+'step_{:02d}.plink.cond'.format(idx),index=None,sep='\\t',header=False)\n",
    "    #pd.Series(conditional_omnibus_list).to_csv(data_out_assoc_phenotype_path+'step_{:02d}.omnibus.cond'.format(idx))\n",
    "    \n",
    "    pd.Series(conditional_omnibus_list).to_csv(data_out_assoc_phenotype_path+'step_{:02d}.omnibus.cond'.format(idx),index=None,sep='\\t',header=False)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "    #Run plink association test\n",
    "    log.info(\"#########################################  STEP {}-2 Plink Association  #########################################\".format(idx))\n",
    "    command='plink2 --bfile {bfile} {assoc_mode} --pheno {pheno} {covar_on} {covar} --condition-list {condition_list} --out {out} --covar-variance-standardize --threads 40'.format(bfile=plink_path,\n",
    "                                                                                                                        assoc_mode='--logistic' if phenotype_type=='binary' else '--linear',\n",
    "                                                                                                                         pheno=data_out_pheno_path+phenotype_name+'.phe',\n",
    "                                                                                                                         covar_on='--covar' if len(covariate_plink.columns[1:])!=0 else '',\n",
    "                                                                                                                         covar=data_out_assoc_phenotype_path+'step_{:02d}.plink.covar'.format(idx) if len(covariate_plink.columns[1:])!=0 else '',\n",
    "                                                                                                                         condition_list=data_out_assoc_phenotype_path+'step_{:02d}.plink.cond'.format(idx),\n",
    "                                                                                                                         out=data_out_assoc_phenotype_path+'step_{:02d}.plink'.format(idx)                                                                         \n",
    "                                                                                                                        )\n",
    "    log.info(command)\n",
    "    stdout,stderr=run_subprocess(command,dry=False)\n",
    "    log.info(stdout)\n",
    "    log.error(stderr)\n",
    "    \n",
    "\n",
    "    \n",
    "    log.info(\"#########################################  STEP {}-3 Omnibus Association  #########################################\".format(idx))\n",
    "    #Run omnibus association test\n",
    "    command='python omnibus_test.py --assoc {assoc_mode} --fam {fam} --aa {aa} --pheno {pheno} {covar_on} {covar} --condition-list {condition_list} --out {out}'.format(assoc_mode='logistic' if phenotype_type=='binary' else 'linear',\n",
    "                                                                                                                         aa=data_out_assoc_phenotype_path+'step_{:02d}.aa'.format(idx),\n",
    "                                                                                                                         fam=plink_path+'.fam',\n",
    "                                                                                                                         pheno=data_out_pheno_path+phenotype_name+'.phe',\n",
    "                                                                                                                         covar_on='--covar' if len(covariate_omnibus.columns[1:])!=0 else '',\n",
    "                                                                                                                         covar=data_out_assoc_phenotype_path+'step_{:02d}.omnibus.covar'.format(idx) if len(covariate_omnibus.columns[1:])!=0 else '',\n",
    "                                                                                                                         condition_list=data_out_assoc_phenotype_path+'step_{:02d}.omnibus.cond'.format(idx),\n",
    "                                                                                                                         out=data_out_assoc_phenotype_path+'step_{:02d}.omnibus'.format(idx)\n",
    "                                                                                                                        )\n",
    "    log.info(command)\n",
    "    stdout,stderr=run_subprocess(command,dry=False)\n",
    "    log.info(stdout)\n",
    "    log.error(stderr)    \n",
    "    \n",
    "    \"\"\"\n",
    "    log.info(\"#########################################  STEP {}-3 Omnibus Association  #########################################\".format(idx))\n",
    "    #Run omnibus association test\n",
    "    command='Rscript OmnibusTest_BHv5_modified.R {assoc_mode} {output} {fam} {aa} {phe} {covar} {covarname} {condvar}'.format(assoc_mode='logistic' if phenotype_type=='binary' else 'linear',\n",
    "                                                                                                                         output=data_out_assoc_phenotype_path+'step_{:02d}.omnibus'.format(idx),\n",
    "                                                                                                                         aa=data_out_assoc_phenotype_path+'step_{:02d}.aa'.format(idx),\n",
    "                                                                                                                         fam=plink_path+'.fam',\n",
    "                                                                                                                         phe=data_out_assoc_phenotype_path+'phenotype.phe',\n",
    "                                                                                                                         covar=data_out_assoc_phenotype_path+'step_{:02d}.omnibus.covar.old'.format(idx) if len(covariate_omnibus.columns[1:])!=0 else 'NA',\n",
    "                                                                                                                         covarname=','.join(covariate_omnibus_old.columns[1:]) if len(covariate_omnibus_old.columns[1:])!=0 else 'NA',\n",
    "                                                                                                                         condvar=','.join(conditional_omnibus_list+conditional_plink_phased_list) if len(conditional_omnibus_list+conditional_plink_phased_list)!=0 else 'NA'\n",
    "                                                                                                                        )\n",
    "    log.info(command)\n",
    "    stdout,stderr=run_subprocess(command,dry=False)\n",
    "    log.info(stdout)\n",
    "    log.error(stderr)\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    log.info(\"#########################################  STEP {}-4 Result Save  #########################################\".format(idx))\n",
    "    \n",
    "    #Data save\n",
    "    #plink_assoc_result=parse_plink_assoc(data_out_assoc_phenotype_path+'step_{:02d}.plink.{}assoc'.format(idx,'' if phenotype_type=='binary' else 'q'))\n",
    "    #plink_assoc_result.to_csv(data_out_assoc_phenotype_path+'step_{:02d}.plink.{}assoc.munged'.format(idx,'' if phenotype_type=='binary' else 'q'))\n",
    "    #plink_assoc_result_sorted=plink_assoc_result.sort_values('P')    \n",
    "    #plink_assoc_result_sorted.to_csv(data_out_assoc_phenotype_path+'step_{:02d}.plink.{}assoc.sorted_'.format(idx,'' if phenotype_type=='binary' else 'q'),sep='\\t',index=None)    \n",
    "    \n",
    "    plink_assoc_result=pd.read_csv(data_out_assoc_phenotype_path+'step_{:02d}.plink.PHENO2.glm.{}'.format(idx,'logistic' if phenotype_type=='binary' else 'linear'),sep='\\t').astype({'P':float})\n",
    "    plink_assoc_result=plink_assoc_result[plink_assoc_result['TEST']=='ADD']\n",
    "    plink_assoc_result.to_csv(data_out_assoc_phenotype_path+'step_{:02d}.plink.PHENO2.glm.{}.munged'.format(idx,'logistic' if phenotype_type=='binary' else 'linear'),sep='\\t',index=None)\n",
    "    plink_assoc_result_sorted=plink_assoc_result.sort_values('P')    \n",
    "    plink_assoc_result_sorted.to_csv(data_out_assoc_phenotype_path+'step_{:02d}.plink.PHENO2.glm.{}.sorted'.format(idx,'logistic' if phenotype_type=='binary' else 'linear'),sep='\\t',index=None)\n",
    "    \n",
    "    omnibus_assoc_result=pd.read_csv(data_out_assoc_phenotype_path+'step_{:02d}.omnibus.{}'.format(idx,'logistic' if phenotype_type=='binary' else 'linear'),sep='\\t').astype({'P':float})\n",
    "    omnibus_assoc_result.to_csv(data_out_assoc_phenotype_path+'step_{:02d}.omnibus.{}.munged'.format(idx,'logistic' if phenotype_type=='binary' else 'linear'),sep='\\t',index=None)\n",
    "    omnibus_assoc_result_sorted=omnibus_assoc_result.sort_values('P')\n",
    "    omnibus_assoc_result_sorted.to_csv(data_out_assoc_phenotype_path+'step_{:02d}.omnibus.{}.sorted'.format(idx,'logistic' if phenotype_type=='binary' else 'linear'),sep='\\t',index=None)\n",
    "    \n",
    "    #Identify Peak variant\n",
    "    log.info(\"Plink association result\")\n",
    "    log.info(plink_assoc_result_sorted.head(5))\n",
    "    log.info(\"Omnibus association result\")\n",
    "    log.info(omnibus_assoc_result_sorted.head(5))\n",
    "    \n",
    "    if plink_assoc_result_sorted.iloc[0]['P'] < omnibus_assoc_result_sorted.iloc[0]['P']:\n",
    "        variant_name=plink_assoc_result_sorted.iloc[0]['ID']\n",
    "        log.info('Plink win!!! {}'.format(variant_name))\n",
    "        conditional_variant_list.append((\"plink\",variant_name))        \n",
    "        if plink_assoc_result_sorted.iloc[0]['P']>5e-8:\n",
    "            log.info('p-value of {}({}) is larger than 5x10^-8. finish conditional analysis'.format(variant_name,plink_assoc_result_sorted.iloc[0]['P']))\n",
    "            break\n",
    "    else:\n",
    "        variant_name=omnibus_assoc_result_sorted.iloc[0]['ID']\n",
    "        log.info('Omnibus win!!! {}'.format(variant_name))\n",
    "        conditional_variant_list.append((\"omnibus\",variant_name))\n",
    "        if omnibus_assoc_result_sorted.iloc[0]['P']>5e-8:\n",
    "            log.info('p-value of {}({}) is larger than 5x10^-8. finish conditional analysis'.format(variant_name,omnibus_assoc_result_sorted.iloc[0]['P']))\n",
    "            break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ 4., 11., 12., ..., nan, nan, nan])"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.unique(y_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "covariate_plink_df.columns[check_duplicate(covariate_plink_df,threshold=0.9999,keep=False)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "covariate_plink_df.columns[check_duplicate(covariate_plink_df,threshold=0.9999,keep=False)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 320,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['age', 'sex', 'AS', 'NC', 'HLA_A*02:01', 'HLA_A*02:03', 'HLA_A*02:06',\n",
       "       'HLA_A*26:01', 'HLA_A*26:03', 'HLA_A*30:04', 'HLA_A*31:01',\n",
       "       'HLA_A*33:03', 'AA_A_9_30018537_exon2_F', 'AA_A_9_30018537_exon2_Y',\n",
       "       'AA_A_56_30018678_exon2', 'AA_A_62_30018696_exon2_Q',\n",
       "       'AA_A_70_30018720_exon2', 'AA_A_76_30018738_exon2_A',\n",
       "       'AA_A_77_30018741_exon2_N', 'AA_A_114_30019093_exon3_R',\n",
       "       'AA_A_116_30019099_exon3_D', 'AA_A_152_30019207_exon3_E',\n",
       "       'AA_A_156_30019219_exon3_Q', 'SNPS_A_28_30018337_exon1',\n",
       "       '6:30165273_C/T'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 320,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "covariate_plink_df.columns[~check_duplicate(covariate_plink_df,threshold=0.9,keep=False)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "log.info_head(\"Step {}-5 Save\".format(idx))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "log.info_head(\"Step {}-6 Plink\".format(idx))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "log.info_head(\"Step {}-7 Omnibus\".format(idx))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_duplicate(array,axis=1,threshold=0.9,keep='last'):\n",
    "    array=array.copy()\n",
    "    if isinstance(array, np.ndarray):\n",
    "        pass\n",
    "    elif isinstance(array, pd.core.frame.DataFrame):\n",
    "        array=array.values\n",
    "        \n",
    "    if axis==1:\n",
    "        pass\n",
    "    elif axis==0:\n",
    "        array=array.transpose()\n",
    "    else:\n",
    "        raise NotImplementedError\n",
    "            \n",
    "    dupcol_check=np.repeat(False,array.shape[1])\n",
    "    for i in range(array.shape[1]):\n",
    "        for j in range(i+1,array.shape[1]):\n",
    "            x_i=array[:,i]\n",
    "            x_j=array[:,j]\n",
    "            if np.all(x_i==x_j) or np.corrcoef(x_i,x_j)[0,1]>threshold:\n",
    "                if keep==False:\n",
    "                    dupcol_check[i]=True\n",
    "                    dupcol_check[j]=True\n",
    "                elif keep=='last':   \n",
    "                    if (not dupcol_check[i]) and (not dupcol_check[j]):\n",
    "                        dupcol_check[j]=True\n",
    "                elif keep=='loose_last':\n",
    "                    dupcol_check[j]=True\n",
    "                else:\n",
    "                    raise NotImplementedError\n",
    "    return dupcol_check"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 311,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\u001b[0;31mSignature:\u001b[0m \u001b[0mcheck_duplicate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marray\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mthreshold\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m0.9\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkeep\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'last'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
       "\u001b[0;31mDocstring:\u001b[0m <no docstring>\n",
       "\u001b[0;31mFile:\u001b[0m      /data/ch6845/MHC_phewas_testbench/<ipython-input-307-1ff9a2e80d4c>\n",
       "\u001b[0;31mType:\u001b[0m      function\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "check_duplicate(covariate_plink_df,keep==False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "AA_re_exp=re.compile('AA_(?P<name>[0-9A-Z]*)_(?P<allele>[0-9]*)_.*')\n",
    "for i in plink_KCHIP_HLA_AA_SNP_1000G_bim[plink_KCHIP_HLA_AA_SNP_1000G_bim.index.str.contains('AA_',regex=True)].index:\n",
    "    print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.delete(dosage_array,dosage_array.sum(axis=0).argmax(),axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "conditional_variant_list=[('plink', '6:30165273_C/T'), ('plink', '6:30170628_T/TA'), ('plink', 'rs2523942'), ('plink', '6:29917446_T/G')]\n",
    "r2_list_dict={}\n",
    "\n",
    "for idx in range(5,10+1):\n",
    "    log.info(\"############################################  STEP {}  ############################################\".format(idx))\n",
    "    log.info(\"##########################  STEP {}-1 Prepare input (covar,cond,etc.)  ###################################\".format(idx))\n",
    "    #Initialize *.aa\n",
    "    if os.path.exists('{}'.format(data_out_assoc_phenotype_path+'step_{:02d}.aa'.format(idx))):\n",
    "        command='rm {}'.format(data_out_assoc_phenotype_path+'step_{:02d}*'.format(idx))\n",
    "        log.info(command);os.system(command)\n",
    "    \n",
    "    #command='cp {} {}'.format()\n",
    "    #print(command);os.system(command)\n",
    "    \n",
    "    command='head -n5 {} > {}'.format(aa_path,data_out_assoc_phenotype_path+'step_{:02d}.aa'.format(idx))\n",
    "    log.info(command); assert os.system(command)==0\n",
    "    \n",
    "    command='grep \" AA_\" {} >> {}'.format(aa_path,data_out_assoc_phenotype_path+'step_{:02d}.aa'.format(idx))\n",
    "    log.info(command); assert os.system(command)==0\n",
    "\n",
    "    \n",
    "    #Initialize conditional\n",
    "    conditional_omnibus_list=[]\n",
    "    conditional_plink_list=[]\n",
    "    conditional_plink_phased_list=[]\n",
    "    \n",
    "    #Initialize covariate\n",
    "    #covariate_plink=pd.DataFrame(index=phenotypes.index) \n",
    "    #covariate_omnibus=pd.DataFrame(index=phenotypes.index) \n",
    "    \n",
    "    covariate_plink=pd.read_csv(PC_path,sep='\\t').set_index('ID').loc[fam['IID']]\n",
    "    covariate_plink['age']=phenotypes['age']\n",
    "    covariate_plink['sex']=phenotypes['sex']\n",
    "    covariate_plink['AS']=phenotypes['cohort'].replace(1,1).replace(2,0).replace(3,0)\n",
    "    covariate_plink['CT']=phenotypes['cohort'].replace(1,0).replace(2,1).replace(3,0)\n",
    "    #covariate_plink['NC']=phenotypes['cohort'].replace(1,0).replace(2,0).replace(3,1)\n",
    "    \n",
    "    covariate_omnibus=pd.read_csv(PC_path,sep='\\t').set_index('ID').loc[fam['IID']]\n",
    "    covariate_omnibus['age']=phenotypes['age']\n",
    "    covariate_omnibus['sex']=phenotypes['sex']\n",
    "    covariate_omnibus['AS']=phenotypes['cohort'].replace(1,1).replace(2,0).replace(3,0)\n",
    "    covariate_omnibus['CT']=phenotypes['cohort'].replace(1,0).replace(2,1).replace(3,0)\n",
    "    #covariate_omnibus['NC']=phenotypes['cohort'].replace(1,0).replace(2,0).replace(3,1)    \n",
    "    \n",
    "    log.info(\"##########################  STEP {}-1-1 conditional variant expansion (assign to a HLA gene)  ###################################\".format(idx))\n",
    "    log.info(\"conditional_variant_list:\"+str(conditional_variant_list))\n",
    "    conditional_variant_list_expand=[]\n",
    "    #expand\n",
    "    for assoc_type,conditional_variant in conditional_variant_list:\n",
    "        log.info('expanding {}'.format(conditional_variant))\n",
    "        if assoc_type=='omnibus':\n",
    "            for variant_name in np.unique(['_'.join(variant.split('_')[:3]) for variant in bim_aa[bim_aa.index.str.contains('AA_'+conditional_variant.split('_')[1]+'_')].index.tolist()]):#AA\n",
    "                conditional_variant_list_expand.append((\"omnibus\",variant_name))\n",
    "                log.info('found {}'.format(variant_name))\n",
    "            for variant_name in bim_aa[bim_aa.index.str.contains('HLA_'+conditional_variant.split('_')[1]+'\\*')].index:#HLA\n",
    "                conditional_variant_list_expand.append((\"plink\",variant_name))\n",
    "                log.info('found {}'.format(variant_name))\n",
    "        elif assoc_type=='plink':\n",
    "            if conditional_variant[:3]=='HLA':\n",
    "                for variant_name in np.unique(['_'.join(variant.split('_')[:3]) for variant in bim_aa[bim_aa.index.str.contains('AA_'+conditional_variant[4:].split('*')[0]+'_')].index.tolist()]):#AA\n",
    "                    conditional_variant_list_expand.append((\"omnibus\",variant_name))\n",
    "                    log.info('found {}'.format(variant_name))\n",
    "                for variant_name in bim_aa[bim_aa.index.str.contains('HLA_'+conditional_variant[4:].split('*')[0]+'\\*')].index:#HLA\n",
    "                    conditional_variant_list_expand.append((\"plink\",variant_name))                \n",
    "                    log.info('found {}'.format(variant_name))\n",
    "            else:#SNPs\n",
    "                conditional_variant_list_expand.append((\"plink\",conditional_variant))\n",
    "                log.info('found {}'.format(conditional_variant))\n",
    "                \n",
    "                r2_list=r2_list_dict.get(conditional_variant)\n",
    "                if r2_list is None: \n",
    "                    r2_list=[]\n",
    "                    for idx_bim,(SNP,row) in enumerate(bim_aa.iterrows()):\n",
    "                        r2=pearsonr(plink_aa.get_geno_marker(conditional_variant),plink_aa.get_geno_marker(SNP))[0]**2\n",
    "                        r2_list.append(r2)\n",
    "                    r2_list_dict[conditional_variant]=r2_list\n",
    "                r2_df=pd.DataFrame(r2_list,index=bim_aa.index)    \n",
    "\n",
    "                gene_assign_count=gene_assign[r2_df[0]>0.7].sum(axis=0).iloc[1:]    \n",
    "                log.info(\"found HLA gene with LD \\n \"+str(gene_assign_count))\n",
    "                \n",
    "                for HLA_type in gene_assign_count[gene_assign_count>0].index.tolist():\n",
    "                    for variant_name in np.unique(['_'.join(variant.split('_')[:3]) for variant in bim_aa[bim_aa.index.str.contains('AA_'+HLA_type+'_')].index.tolist()]):#AA\n",
    "                        conditional_variant_list_expand.append((\"omnibus\",variant_name))\n",
    "                        log.info('found {}'.format(variant_name))\n",
    "                    for variant_name in bim_aa[bim_aa.index.str.contains('HLA_'+HLA_type+'\\*')].index:#HLA\n",
    "                        conditional_variant_list_expand.append((\"plink\",variant_name))\n",
    "                        log.info('found {}'.format(variant_name))\n",
    "        else:\n",
    "            raise\n",
    "    log.info(\"conditional_variant_list_expand: \"+str(conditional_variant_list_expand))\n",
    "    log.info(\"##########################  STEP {}-1-2 conditional variant -> omnibus,plink,plink_phased  ###################################\".format(idx))\n",
    "    #Split conditional\n",
    "    for assoc_type,conditional_variant in conditional_variant_list_expand:\n",
    "        if assoc_type=='omnibus':\n",
    "            conditional_omnibus_list.append(conditional_variant)\n",
    "            log.info(\"{} -> omnibus\".format(conditional_variant))\n",
    "        elif assoc_type=='plink':\n",
    "            if os.path.exists('{}temp.aa'.format(data_out_assoc_phenotype_path)):\n",
    "                os.system('rm {}temp.aa'.format(data_out_assoc_phenotype_path))\n",
    "            os.system('grep \"{} \" {} > {}temp.aa'.format(conditional_variant.replace('*','\\*'),aa_path,data_out_assoc_phenotype_path));f=open('{}temp.aa'.format(data_out_assoc_phenotype_path),'r');lines=f.readlines();f.close();os.system('rm {}temp.aa'.format(data_out_assoc_phenotype_path))\n",
    "            if len(lines)==0:\n",
    "                conditional_plink_list.append(conditional_variant)\n",
    "                log.info(\"{} -> plink\".format(conditional_variant))\n",
    "            elif len(lines)==1:\n",
    "                conditional_plink_phased_list.append(conditional_variant)\n",
    "                log.info(\"{} -> plink_phased\".format(conditional_variant))\n",
    "            else:\n",
    "                raise\n",
    "        else:\n",
    "            raise\n",
    "\n",
    "    conditional_omnibus_list=list(np.unique(conditional_omnibus_list))\n",
    "    conditional_plink_list=list(np.unique(conditional_plink_list))\n",
    "    conditional_plink_phased_list=list(np.unique(conditional_plink_phased_list))\n",
    "    \n",
    "    log.info('conditional_omnibus_list: '+str(conditional_omnibus_list))\n",
    "    log.info('conditional_plink_list: '+str(conditional_plink_list))\n",
    "    log.info('conditional_plink_phased_list: '+str(conditional_plink_phased_list))\n",
    "    \n",
    "    \n",
    "    #Process condition\n",
    "    #(1) Omnibus AA peak\n",
    "    log.info(\"##########################  STEP {}-1-3 process conditional_omnibus_list  ###################################\".format(idx))\n",
    "    \n",
    "    \n",
    "    for conditional_variant in conditional_omnibus_list:\n",
    "        log.info(\"Process {}\".format(conditional_variant))\n",
    "        #individual_list=fam.IID#pd.read_csv(plink_path+'.fam',header=None,sep=' ')[1]\n",
    "        if os.path.exists('{}temp.aa'.format(data_out_assoc_phenotype_path)):\n",
    "            os.system('rm {}temp.aa'.format(data_out_assoc_phenotype_path))\n",
    "        os.system('grep \"{} \" {} > {}temp.aa'.format(conditional_variant.replace('*','\\*'),aa_path,data_out_assoc_phenotype_path));f=open('{}temp.aa'.format(data_out_assoc_phenotype_path),'r');lines=f.readlines();f.close();os.system('rm {}temp.aa'.format(data_out_assoc_phenotype_path))\n",
    "        marker_name=lines[0].split(' ')[1]\n",
    "        assert marker_name==conditional_variant;assert len(lines)==1\n",
    "        aa_call_list=np.array(lines[0].strip().split(' ')[2:])\n",
    "        aa_call_list_group=np.array([(aa_call_list[i],aa_call_list[i+1]) for i in range(0,len(aa_call_list),2)])\n",
    "\n",
    "        residue_dosage_all=pd.DataFrame(index=fam.IID)\n",
    "        for residue in np.unique(aa_call_list):\n",
    "            log.info(\"residue {} \".format(residue)+(\"skipped\" if residue=='NA' else \"dosage count\"))\n",
    "            if residue==\"NA\":\n",
    "                continue\n",
    "            residue_dosage=pd.Series(0,index=fam.IID)\n",
    "            residue_dosage.loc[(aa_call_list_group[:,0]==residue)]+=1\n",
    "            residue_dosage.loc[(aa_call_list_group[:,1]==residue)]+=1\n",
    "            residue_dosage_all[residue]=residue_dosage\n",
    "        log.info(\"ignored {} among \\n{}\".format(residue_dosage_all.sum().sort_values(ascending=True).index[-1],residue_dosage_all.sum().sort_values(ascending=True)))\n",
    "\n",
    "        residue_dosage_all_excluded=residue_dosage_all[residue_dosage_all.sum().sort_values(ascending=True)[:-1].index]#exclude the most frequent residue\n",
    "        for residue,row in residue_dosage_all_excluded.items():\n",
    "            covariate_plink[conditional_variant+'_'+residue]=row        \n",
    "        log.info(\"plink covariate \"+str(covariate_plink.columns))    \n",
    "        \n",
    "    #(2) Plink unphased peak (1000G-SNP)\n",
    "    log.info(\"##########################  STEP {}-1-4 process conditional_plink_list  ###################################\".format(idx))\n",
    "    for conditional_variant in conditional_plink_list:\n",
    "        log.info(\"** process {}\".format(conditional_variant))\n",
    "        dosage=plink.get_geno_marker(conditional_variant).astype(float)\n",
    "        dosage[dosage==-1]=np.nan\n",
    "        covariate_omnibus[conditional_variant]=dosage\n",
    "        log.info(\"added dosage of {} to omnibus covariate\".format(conditional_variant))\n",
    "        \n",
    "    #(3) Plink phased peak (KCHIP-HLA, SNP)\n",
    "    log.info(\"##########################  STEP {}-1-5 process conditional_plink_phased_list  ###################################\".format(idx))\n",
    "    for conditional_variant in conditional_plink_phased_list:\n",
    "        log.info(\"** process {}\".format(conditional_variant))\n",
    "        #Attach to aa file\n",
    "        if os.path.exists('{}temp.aa'.format(data_out_assoc_phenotype_path)):\n",
    "            os.system('rm {}temp.aa'.format(data_out_assoc_phenotype_path));\n",
    "        \n",
    "        os.system('grep \"{} \" {} > {}temp.aa'.format(conditional_variant.replace('*','\\*'),aa_path,data_out_assoc_phenotype_path));f=open('{}temp.aa'.format(data_out_assoc_phenotype_path),'r');lines=f.readlines();f.close();assert len(lines)==1\n",
    "        os.system('cat {}temp.aa >> {}'.format(data_out_assoc_phenotype_path,data_out_assoc_phenotype_path+'step_{:02d}.aa'.format(idx)))\n",
    "        os.system('rm {}temp.aa'.format(data_out_assoc_phenotype_path))\n",
    "        log.info(\"added phased information of {} to aa file\".format(conditional_variant))\n",
    "    \n",
    "    \n",
    "    \n",
    "    log.info(\"##########################  STEP {}-1-6 save covariate, condition  ###################################\".format(idx))\n",
    "    log.info('convarite_omnibus: '+str(covariate_omnibus.columns[:] if len(covariate_omnibus.columns)!=0 else []))\n",
    "    log.info('convarite_plink: '+str(covariate_plink.columns[:] if len(covariate_plink.columns)!=0 else []))\n",
    "    \n",
    "    \n",
    "    \n",
    "    covariate_plink=covariate_plink.reset_index()\n",
    "    covariate_plink=covariate_plink.drop(columns=covariate_plink.iloc[:,1:].columns[get_dupbool(covariate_plink.iloc[:,1:])])\n",
    "    if covariate_plink.shape[1]>1:\n",
    "        covariate_plink[[covariate_plink.columns[0],covariate_plink.columns[0]]+covariate_plink.columns[1:].values.tolist()].fillna(-9).to_csv(data_out_assoc_phenotype_path+'step_{:02d}.plink.covar'.format(idx),index=None,sep='\\t')\n",
    "        #covariate_plink[[covariate_plink.columns[0],covariate_plink.columns[0]]].to_csv(data_out_assoc_phenotype_path+'step_{:02d}.plink.covar'.format(idx),index=None,sep='\\t')\n",
    "    \n",
    "    covariate_omnibus=covariate_omnibus.reset_index()\n",
    "    covariate_omnibus=covariate_omnibus.drop(columns=covariate_omnibus.iloc[:,1:].columns[get_dupbool(covariate_omnibus.iloc[:,1:])])\n",
    "    if covariate_omnibus.shape[1]>1:\n",
    "        covariate_omnibus[[covariate_omnibus.columns[0],covariate_omnibus.columns[0]]+covariate_omnibus.columns[1:].values.tolist()].fillna(-9).to_csv(data_out_assoc_phenotype_path+'step_{:02d}.omnibus.covar'.format(idx),index=None,sep='\\t')\n",
    "        \n",
    "    \"\"\"\n",
    "    ##FOR OLD\n",
    "    covariate_omnibus_old=covariate_omnibus.copy()\n",
    "    covariate_omnibus_old.columns=['header{}'.format(i) for i in range(len(covariate_omnibus_old.columns))]\n",
    "    covariate_omnibus_old=covariate_omnibus_old.reset_index()\n",
    "    if covariate_omnibus_old.shape[1]>1:\n",
    "        covariate_omnibus_old[covariate_omnibus_old.columns[1:]].to_csv(data_out_assoc_phenotype_path+'step_{:02d}.omnibus.covar.old'.format(idx),index=None,sep='\\t',na_rep='NA')\n",
    "        #covariate_omnibus[[covariate_omnibus.columns[0]]].to_csv(data_out_assoc_phenotype_path+'step_{:02d}.omnibus.covar'.format(idx),index=None,sep='\\t')\n",
    "    \"\"\"\n",
    "    \n",
    "    conditional_plink=conditional_plink_list+conditional_plink_phased_list\n",
    "    dosage_array=np.array([np.where(plink.get_geno_marker(conditional_variant)==-1, np.nan, plink.get_geno_marker(conditional_variant)) for conditional_variant in conditional_plink])\n",
    "    pd.Series(conditional_plink)[get_dupbool(dosage_array,axis=0)].to_csv(data_out_assoc_phenotype_path+'step_{:02d}.plink.cond'.format(idx),index=None,sep='\\t',header=False)\n",
    "    \n",
    "    dosage_list=[]\n",
    "    for conditional_variant in conditional_plink:\n",
    "        doage=plink_aa.get_geno_marker(conditional_variant)\n",
    "        dosage[dosage==-1]=np.nan\n",
    "        dosage_list.append(dosage)\n",
    "    dosage_list=np.array([])\n",
    "    \n",
    "    \n",
    "    conditional_omnibus=cconditional_omnibus_list+conditional_plink_phased_list\n",
    "    np.array([np.where(plink.get_geno_marker(conditional_variant)==-1, np.nan, plink.get_geno_marker(conditional_variant)) for conditional_variant in conditional_plink_list+conditional_plink_phased_list])\n",
    "    \n",
    "    dosage_list=[]\n",
    "    for conditional_variant in conditional_plink_list+conditional_plink_phased_list:\n",
    "        print(conditional_variant)\n",
    "        doage=plink.get_geno_marker(conditional_variant)\n",
    "        dosage[dosage==-1]=np.nan\n",
    "        dosage_list.append(dosage)\n",
    "    dosage_list=np.array(dosage_list) \n",
    "\n",
    "\n",
    "    dupcol_find(dosage_list.transpose())\n",
    "\n",
    "    \n",
    "    \n",
    "\n",
    "    \n",
    "    \n",
    "    \n",
    "    pd.Series(conditional_plink_list+conditional_plink_phased_list).to_csv(data_out_assoc_phenotype_path+'step_{:02d}.plink.cond'.format(idx),index=None,sep='\\t',header=False)\n",
    "    #pd.Series(conditional_omnibus_list).to_csv(data_out_assoc_phenotype_path+'step_{:02d}.omnibus.cond'.format(idx))\n",
    "    \n",
    "    pd.Series(conditional_omnibus_list+conditional_plink_phased_list).to_csv(data_out_assoc_phenotype_path+'step_{:02d}.omnibus.cond'.format(idx),index=None,sep='\\t',header=False)\n",
    "    \n",
    "    #Run plink association test\n",
    "    log.info(\"#########################################  STEP {}-2 Plink Association  #########################################\".format(idx))\n",
    "    command='plink2 --bfile {bfile} {assoc_mode} --pheno {pheno} {covar_on} {covar} --condition-list {condition_list} --out {out} --covar-variance-standardize --threads 40'.format(bfile=plink_path,\n",
    "                                                                                                                        assoc_mode='--logistic' if phenotype_type=='binary' else '--linear',\n",
    "                                                                                                                         pheno=data_out_pheno_path+phenotype_name+'.phe',\n",
    "                                                                                                                         covar_on='--covar' if len(covariate_plink.columns[1:])!=0 else '',\n",
    "                                                                                                                         covar=data_out_assoc_phenotype_path+'step_{:02d}.plink.covar'.format(idx) if len(covariate_plink.columns[1:])!=0 else '',\n",
    "                                                                                                                         condition_list=data_out_assoc_phenotype_path+'step_{:02d}.plink.cond'.format(idx),\n",
    "                                                                                                                         out=data_out_assoc_phenotype_path+'step_{:02d}.plink'.format(idx)                                                                         \n",
    "                                                                                                                        )\n",
    "    log.info(command)\n",
    "    stdout,stderr=run_subprocess(command,dry=False)\n",
    "    log.info(stdout)\n",
    "    log.error(stderr)\n",
    "    \n",
    "\n",
    "    \n",
    "    log.info(\"#########################################  STEP {}-3 Omnibus Association  #########################################\".format(idx))\n",
    "    #Run omnibus association test\n",
    "    command='python omnibus_test.py --assoc {assoc_mode} --fam {fam} --aa {aa} --pheno {pheno} {covar_on} {covar} --condition-list {condition_list} --out {out}'.format(assoc_mode='logistic' if phenotype_type=='binary' else 'linear',\n",
    "                                                                                                                         aa=data_out_assoc_phenotype_path+'step_{:02d}.aa'.format(idx),\n",
    "                                                                                                                         fam=plink_path+'.fam',\n",
    "                                                                                                                         pheno=data_out_pheno_path+phenotype_name+'.phe',\n",
    "                                                                                                                         covar_on='--covar' if len(covariate_omnibus.columns[1:])!=0 else '',\n",
    "                                                                                                                         covar=data_out_assoc_phenotype_path+'step_{:02d}.omnibus.covar'.format(idx) if len(covariate_omnibus.columns[1:])!=0 else '',\n",
    "                                                                                                                         condition_list=data_out_assoc_phenotype_path+'step_{:02d}.omnibus.cond'.format(idx),\n",
    "                                                                                                                         out=data_out_assoc_phenotype_path+'step_{:02d}.omnibus'.format(idx)\n",
    "                                                                                                                        )\n",
    "    log.info(command)\n",
    "    stdout,stderr=run_subprocess(command,dry=False)\n",
    "    log.info(stdout)\n",
    "    log.error(stderr)    \n",
    "    \n",
    "    \"\"\"\n",
    "    log.info(\"#########################################  STEP {}-3 Omnibus Association  #########################################\".format(idx))\n",
    "    #Run omnibus association test\n",
    "    command='Rscript OmnibusTest_BHv5_modified.R {assoc_mode} {output} {fam} {aa} {phe} {covar} {covarname} {condvar}'.format(assoc_mode='logistic' if phenotype_type=='binary' else 'linear',\n",
    "                                                                                                                         output=data_out_assoc_phenotype_path+'step_{:02d}.omnibus'.format(idx),\n",
    "                                                                                                                         aa=data_out_assoc_phenotype_path+'step_{:02d}.aa'.format(idx),\n",
    "                                                                                                                         fam=plink_path+'.fam',\n",
    "                                                                                                                         phe=data_out_assoc_phenotype_path+'phenotype.phe',\n",
    "                                                                                                                         covar=data_out_assoc_phenotype_path+'step_{:02d}.omnibus.covar.old'.format(idx) if len(covariate_omnibus.columns[1:])!=0 else 'NA',\n",
    "                                                                                                                         covarname=','.join(covariate_omnibus_old.columns[1:]) if len(covariate_omnibus_old.columns[1:])!=0 else 'NA',\n",
    "                                                                                                                         condvar=','.join(conditional_omnibus_list+conditional_plink_phased_list) if len(conditional_omnibus_list+conditional_plink_phased_list)!=0 else 'NA'\n",
    "                                                                                                                        )\n",
    "    log.info(command)\n",
    "    stdout,stderr=run_subprocess(command,dry=False)\n",
    "    log.info(stdout)\n",
    "    log.error(stderr)\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    log.info(\"#########################################  STEP {}-4 Result Save  #########################################\".format(idx))\n",
    "    \n",
    "    #Data save\n",
    "    #plink_assoc_result=parse_plink_assoc(data_out_assoc_phenotype_path+'step_{:02d}.plink.{}assoc'.format(idx,'' if phenotype_type=='binary' else 'q'))\n",
    "    #plink_assoc_result.to_csv(data_out_assoc_phenotype_path+'step_{:02d}.plink.{}assoc.munged'.format(idx,'' if phenotype_type=='binary' else 'q'))\n",
    "    #plink_assoc_result_sorted=plink_assoc_result.sort_values('P')    \n",
    "    #plink_assoc_result_sorted.to_csv(data_out_assoc_phenotype_path+'step_{:02d}.plink.{}assoc.sorted_'.format(idx,'' if phenotype_type=='binary' else 'q'),sep='\\t',index=None)    \n",
    "    \n",
    "    plink_assoc_result=pd.read_csv(data_out_assoc_phenotype_path+'step_{:02d}.plink.PHENO2.glm.{}'.format(idx,'logistic' if phenotype_type=='binary' else 'linear'),sep='\\t').astype({'P':float})\n",
    "    plink_assoc_result=plink_assoc_result[plink_assoc_result['TEST']=='ADD']\n",
    "    plink_assoc_result.to_csv(data_out_assoc_phenotype_path+'step_{:02d}.plink.PHENO2.glm.{}.munged'.format(idx,'logistic' if phenotype_type=='binary' else 'linear'),sep='\\t',index=None)\n",
    "    plink_assoc_result_sorted=plink_assoc_result.sort_values('P')    \n",
    "    plink_assoc_result_sorted.to_csv(data_out_assoc_phenotype_path+'step_{:02d}.plink.PHENO2.glm.{}.sorted'.format(idx,'logistic' if phenotype_type=='binary' else 'linear'),sep='\\t',index=None)\n",
    "    \n",
    "    omnibus_assoc_result=pd.read_csv(data_out_assoc_phenotype_path+'step_{:02d}.omnibus.{}'.format(idx,'logistic' if phenotype_type=='binary' else 'linear'),sep='\\t').astype({'P':float})\n",
    "    omnibus_assoc_result.to_csv(data_out_assoc_phenotype_path+'step_{:02d}.omnibus.{}.munged'.format(idx,'logistic' if phenotype_type=='binary' else 'linear'),sep='\\t',index=None)\n",
    "    omnibus_assoc_result_sorted=omnibus_assoc_result.sort_values('P')\n",
    "    omnibus_assoc_result_sorted.to_csv(data_out_assoc_phenotype_path+'step_{:02d}.omnibus.{}.sorted'.format(idx,'logistic' if phenotype_type=='binary' else 'linear'),sep='\\t',index=None)\n",
    "    \n",
    "    #Identify Peak variant\n",
    "    log.info(\"Plink association result\")\n",
    "    log.info(plink_assoc_result_sorted.head(5))\n",
    "    log.info(\"Omnibus association result\")\n",
    "    log.info(omnibus_assoc_result_sorted.head(5))\n",
    "    \n",
    "    if plink_assoc_result_sorted.iloc[0]['P'] < omnibus_assoc_result_sorted.iloc[0]['P']:\n",
    "        variant_name=plink_assoc_result_sorted.iloc[0]['ID']\n",
    "        log.info('Plink win!!! {}'.format(variant_name))\n",
    "        conditional_variant_list.append((\"plink\",variant_name))        \n",
    "        if plink_assoc_result_sorted.iloc[0]['P']>5e-8:\n",
    "            log.info('p-value of {}({}) is larger than 5x10^-8. finish conditional analysis'.format(variant_name,plink_assoc_result_sorted.iloc[0]['P']))\n",
    "            break\n",
    "    else:\n",
    "        variant_name=omnibus_assoc_result_sorted.iloc[0]['ID']\n",
    "        log.info('Omnibus win!!! {}'.format(variant_name))\n",
    "        conditional_variant_list.append((\"omnibus\",variant_name))\n",
    "        if omnibus_assoc_result_sorted.iloc[0]['P']>5e-8:\n",
    "            log.info('p-value of {}({}) is larger than 5x10^-8. finish conditional analysis'.format(variant_name,omnibus_assoc_result_sorted.iloc[0]['P']))\n",
    "            break\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.Series(conditional_plink_list+conditional_plink_phased_list).to_csv(data_out_assoc_phenotype_path+'step_{:02d}.plink.cond'.format(idx),index=None,sep='\\t',header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.Series(conditional_plink_list+conditional_plink_phased_list)[dupcol_find(dosage_list.transpose())].to_csv(data_out_assoc_phenotype_path+'step_{:02d}.plink.cond'.format(idx),index=None,sep='\\t',header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_dupbool??"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_dupbool(array,axis=1,threshold=0.9,keep='last'):\n",
    "    array=array.copy()\n",
    "    if isinstance(array, np.ndarray):\n",
    "        pass\n",
    "    elif isinstance(array, pd.core.frame.DataFrame):\n",
    "        array=array.values\n",
    "        \n",
    "    if axis==1:\n",
    "        pass\n",
    "    elif axis==0:\n",
    "        array=array.transpose()\n",
    "    else:\n",
    "        raise NotImplementedError\n",
    "            \n",
    "    dupcol_check=np.repeat(False,array.shape[1])\n",
    "    for i in range(array.shape[1]):\n",
    "        for j in range(i+1,array.shape[1]):\n",
    "            x_i=array[:,i]\n",
    "            x_j=array[:,j]\n",
    "            if i==8 or j==8:\n",
    "                print(i,j,np.corrcoef(x_i,x_j)[0,1])\n",
    "            if np.all(x_i==x_j) or np.abs(np.corrcoef(x_i,x_j)[0,1])>threshold:\n",
    "                if keep==False:\n",
    "                    dupcol_check[i]=True\n",
    "                    dupcol_check[j]=True\n",
    "                elif keep=='last':   \n",
    "                    if (not dupcol_check[i]) and (not dupcol_check[j]):\n",
    "                        dupcol_check[j]=True\n",
    "                elif keep=='loose_last':\n",
    "                    dupcol_check[j]=True\n",
    "                else:\n",
    "                    raise NotImplementedError\n",
    "    return dupcol_check"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "conditional_plink=conditional_plink_list+conditional_plink_phased_list\n",
    "pd.Series(conditional_plink)[~get_dupbool(np.array([plink.get_geno_marker(conditional_variant) for conditional_variant in conditional_plink]),axis=0,threshold=0.0001,keep=False)].to_csv(data_out_assoc_phenotype_path+'step_{:02d}.plink.cond'.format(idx),index=None,sep='\\t',header=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "[~get_dupbool(np.array([plink.get_geno_marker(conditional_variant) for conditional_variant in conditional_plink]),axis=0,threshold=0.7,keep='last')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "[~get_dupbool(np.array([plink.get_geno_marker(conditional_variant) for conditional_variant in conditional_plink]),axis=0,threshold=0.000001,keep=False)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!plink2 --bfile data/genotype/4_merge/KCHIP_HLA_SNP_1000G_merged --linear --pheno data/out_pheno/ALP.phe --covar data/out_assoc/ALP/step_05.plink.covar --condition-list data/out_assoc/ALP/step_05.plink.cond --out data/out_assoc/ALP/step_05.plink --covar-variance-standardize --threads 40"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "covariate_plink=covariate_plink.drop(columns=covariate_plink.columns[1:][get_dupbool(covariate_plink.iloc[:,1:],threshold=0.5)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if covariate_plink.shape[1]>1:\n",
    "    covariate_plink[[covariate_plink.columns[0],covariate_plink.columns[0]]+covariate_plink.columns[1:].values.tolist()].fillna(-9).to_csv(data_out_assoc_phenotype_path+'step_{:02d}.plink.covar'.format(idx),index=None,sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "covariate_plink.iloc[:,1:][:,]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_dupbool(covariate_plink.iloc[:,1:],threshold=0.7)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "covariate_plink.iloc[:,1:].columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if covariate_plink.shape[1]>1:\n",
    "    covariate_plink[[covariate_plink.columns[0],covariate_plink.columns[0]]+covariate_plink.columns[1:].values.tolist()].fillna(-9).to_csv(data_out_assoc_phenotype_path+'step_{:02d}.plink.covar'.format(idx),index=None,sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_dupbool(covariate_plink.iloc[:,1:],threshold=0.7,keep=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "covariate_plink.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dupcol_list=dupcol_find(covariate_plink.iloc[:,1:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dupcol_list=dupcol_find(covariate_plink.iloc[:,1:])\n",
    "\n",
    "covariate_plink=covariate_plink.drop(columns=dupcol_list)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in range(1,len(covariate_plink.columns)):\n",
    "    print(covariate_plink.columns[i],covariate_plink.columns[i+1])\n",
    "    print(np.unique(covariate_plink.iloc[:,i]+covariate_plink.iloc[:,i+1]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "dupcol_find(covariate_plink.iloc[:,1:])    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#np.linalg.matrix_rank(x_data_covar)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "corr=covariate_plink.corr()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.linalg.matrix_rank(corr.values[:,:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "covariate_plink.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "a=np.array([plink.get_geno_marker(conditional_variant) for conditional_variant in conditional_plink_list+conditional_plink_phased_list])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_dupbool(a,axis=0,threshold=0.9,keep=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "idx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "a.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "[plink.get_geno_marker(conditional_variant) for conditional_variant in (conditional_plink_list+conditional_plink_phased_list)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dosage_list=[]\n",
    "for conditional_variant in conditional_plink_list+conditional_plink_phased_list:\n",
    "    #print(conditional_variant)\n",
    "    dosage=plink.get_geno_marker(conditional_variant)\n",
    "    #dosage[dosage==-1]=np.nan\n",
    "    dosage_list.append(dosage)\n",
    "dosage_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dosage_list=np.array(dosage_list) \n",
    "dosage_list.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def dupcol_find(array,axis=1,threshold=0.99):\n",
    "    array=array.copy()\n",
    "    if isinstance(array, np.ndarray):\n",
    "        pass\n",
    "    elif isinstance(array, pd.core.frame.DataFrame):\n",
    "        array=array.values\n",
    "        \n",
    "    if axis==1:\n",
    "        pass\n",
    "    elif axis==0:\n",
    "        array=array.transpose()\n",
    "    else:\n",
    "        raise NotImplementedError\n",
    "            \n",
    "    dupcol_check=np.repeat(False,array.shape[1])\n",
    "    for i in range(array.shape[1]):\n",
    "        for j in range(i+1,array.shape[1]):\n",
    "            if (not dupcol_check[i]) and (not dupcol_check[j]):\n",
    "                x_i=array[:,i]\n",
    "                x_j=array[:,j]\n",
    "                if np.all(x_i==x_j) or np.corrcoef(x_i,x_j)[0,1]>threshold:\n",
    "                    dupcol_check[j]=True\n",
    "    return dupcol_check"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
